---
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  rmdformats::readthedown:
    toc_depth: 3
    code_folding: hide
    number_sections: true
---

```{css, echo=FALSE}
#content{
    max-width:1920px;
}
```

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
library(knitr)
library(tibble)
library(dplyr)
library(stringr)
library(tidyr)

knitr::opts_chunk$set(
  echo=TRUE, 
  include=TRUE, 
  warning=FALSE, 
  message=FALSE, 
  results="asis"
)
```

```{r}
source('reportFunctions.R')

meth_files=fread('fileList1.abismal.txt', header=FALSE, data.table=FALSE)

myoptions = read_file_map('fileList2.abismal.txt', do_unlist=FALSE)
task_name=myoptions$task_name
email=myoptions$email
affiliation=get_affiliation(myoptions)

x_axis_angle = ifelse(is.null(myoptions$x_axis_angle), 45, as.numeric(myoptions$x_axis_angle))

get_bar_width_no_legend = function(nsample) {
  return (max(5, 0.25 * nsample))
}

get_bar_width_with_legend = function(nsample) {
  get_bar_width_no_legend(nsample) + 1.5
}

mytheme = function(...) {
  theme(plot.title = element_text(hjust = 0.5),
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = x_axis_angle, hjust = 1),
        ...)
}

bar_height = 4
```

---
title: "Abismal Summary - `r task_name`"
author:
- name: `r email`
  affiliation: `r affiliation`
---

# Reads mapped to genome

```{r}
mapstats_files = meth_files |> dplyr::filter(grepl("mapstats$", V1))

mapstats_lst = apply(mapstats_files, 1, function(row) {
  file_path = row[1]  
  sample_name = row[2]
  df = fread(file_path, header=FALSE, sep=":", data.table=FALSE) |>
    dplyr::rename("Metric"="V1", "Value"="V2") |>
    dplyr::mutate(Sample=sample_name) 
  return(df)
})
mapstats_df = do.call(rbind, mapstats_lst) 
```

```{r fig.cap=figRef("reads_mapped_summary", "Read mapping summary")}
mapped_levels = c("num_unique", "num_ambiguous", "num_unmapped" )
mapped = mapstats_df |> dplyr::filter(Metric %in% mapped_levels) |>
  dplyr::mutate(Metric = factor(Metric, levels=mapped_levels))

g = ggplot(mapped, aes(x=Sample, y=as.numeric(Value), fill=Metric)) + 
  geom_bar(stat="identity", position = "stack", width=0.5) +
  theme_classic() +
  labs(y="Number of reads", title="Read mapping summary") + 
  mytheme()

width=get_bar_width_with_legend(length(unique(mapped$Sample)))
reads_mapped_bar_png=paste0(task_name, ".reads_mapped.bar.png")
ggsave(reads_mapped_bar_png, g, width=width, height=bar_height, units="in", dpi=300, bg="white")
include_graphics(reads_mapped_bar_png)
```

# Read off bait

Off-bait reads represent the reads that fall outside of the targeted bait regions in the sequencing experiment.

```{r}
hsmetrics_files = meth_files |> dplyr::filter(grepl("hs_metrics", V1))

hsmetrics_lst = apply(hsmetrics_files, 1, function(row) {
  file_path = row[1]  
  sample_name = row[2]
  df = fread(file_path, data.table=FALSE) |>
    dplyr::mutate(Sample=sample_name) 
  return(df)
})
hsmetrics_df = do.call(rbind, hsmetrics_lst) 
```

```{r fig.cap=figRef("reads_off_bait_summary", "Reads off bait summary")}
g = ggplot(hsmetrics_df, aes(x=Sample, y=as.numeric(PCT_OFF_BAIT))) + 
  geom_bar(stat="identity", width=0.5) +
  theme_classic() +
  scale_y_continuous(labels = scales::percent) +  
  labs(y="PCT_OFF_BAIT", title="Reads off bait") + 
  mytheme()

width=get_bar_width_no_legend(length(unique(hsmetrics_df$Sample)))
reads_off_bait_bar_png=paste0(task_name, ".reads_off_bait.bar.png")
ggsave(reads_off_bait_bar_png, g, width=width, height=bar_height, units="in", dpi=300, bg="white")
include_graphics(reads_off_bait_bar_png)
```

# Duplication rate

The duplicate reads were detected by dnmtools uniq command. 
The duplication rate indicates the proportion of duplicate reads in the sequencing data.

```{r dup_summary}
dupstats_files = meth_files |> dplyr::filter(grepl(".bam.dupstats$", V1))

dupstats_lst = apply(dupstats_files, 1, function(row) {
  file_path = row[1]  
  sample_name = row[2]
  df = fread(file_path, sep=" ", header=FALSE, data.table=FALSE) |>
    mutate(V1 = str_replace(V1, ":", "")) |>
    dplyr::mutate(Sample=sample_name) 
  return(df)
})
dup_df = do.call(rbind, dupstats_lst)

write.csv(file=paste0(task_name, ".dupstats.csv"), dup_df)

dup_rate <- dup_df |>
  dplyr::filter(V1=="duplication_rate") |>
  mutate(V2 = as.numeric(V2) / 100.0)

dup_width=get_bar_width_no_legend(nrow(dup_rate$Sample))

dup_png=paste0(task_name, ".duplication_rate.png")
ymax=min(1.0, ceiling(max(dup_rate$V2) * 10) / 10)

g3<-ggplot(dup_rate, aes(x=Sample, y=V2)) +
  geom_bar(stat="identity", width = 0.5) +
  ylab("Duplication Rate") +
  scale_y_continuous(labels = scales::percent) +  
  theme_classic() +
  mytheme()
ggsave(dup_png, g3, width=dup_width, height=bar_height, units="in", dpi=300, bg="white")
```

```{r fig.cap=figRef("duplication_rate_summary", "Duplication rate summary")}
include_graphics(dup_png)
```

# Final unique reads

After removing duplicates and reads out of the targeted bait regions, 
the final unique reads represent the high-confidence reads that are used for downstream analysis.

```{r}
flagstat_files = meth_files |> dplyr::filter(grepl(".bam.flagstat", V1))

flagstat_lst = apply(flagstat_files, 1, function(row) {
  file_path = row[1]  
  sample_name = row[2]
  df = fread(file_path, sep=" ", header=FALSE, data.table=FALSE) |>
    dplyr::mutate(Sample=sample_name) 
  return(df)
})
flagstat_df = do.call(rbind, flagstat_lst) |>
  dplyr::filter(V4=="primary")
```

```{r fig.cap=figRef("final_unique_reads_summary", "Final unique reads summary")}
g = ggplot(flagstat_df, aes(x=Sample, y=as.numeric(V1))) + 
  geom_bar(stat="identity", width=0.5) +
  theme_classic() +
  labs(y="Unique reads", title="Final unique reads") + 
  mytheme()

width=get_bar_width_no_legend(length(unique(flagstat_df$Sample)))
final_unique_reads_bar_png=paste0(task_name, ".final_unique_reads.bar.png")
ggsave(final_unique_reads_bar_png, g, width=width, height=bar_height, units="in", dpi=300, bg="white")
include_graphics(final_unique_reads_bar_png)
```

# The RRBS summary metrics

The metrics from reduced representation bisulfite sequencing (Rrbs) data was performed using the Picard CollectRrbsMetrics tool.

## RRBS summary metrics table

The RRBS summary metrics provide insights into the efficiency of bisulfite conversion during the RRBS library preparation process.

```{r}
rrbs_summary_metrics_files = meth_files |> dplyr::filter(grepl(".rrbs_summary_metrics$", V1))

rrbs_summary_metrics_lst = apply(rrbs_summary_metrics_files, 1, function(row) {
  file_path = row[1]  
  sample_name = row[2]
  df = read.delim(file_path, header = T, sep = "\t", stringsAsFactors = FALSE, comment.char = "#", blank.lines.skip = TRUE, nrows = 1) |>
    dplyr::mutate(Sample=sample_name) 
  return(df)
})
rrbs_summary_metrics_df = do.call(rbind, rrbs_summary_metrics_lst) |>
  dplyr::select(Sample, everything()) |>
  dplyr::select(-SAMPLE, -LIBRARY, -READ_GROUP)

write.csv(file=paste0(task_name, ".rrbs_summary_metrics.csv"), rrbs_summary_metrics_df, row.names=FALSE)
print_table(rrbs_summary_metrics_df, byDT=TRUE, row.names=FALSE)
```

## The BS rates of Non-CpG and CpG sites

- **Non-CpG sites**: Measures the efficiency of bisulfite conversion at cytosines outside of CpG dinucleotides. High conversion rates (>95%) indicate successful bisulfite treatment.
- **CpG sites**: Measures the bisulfite conversion at cytosines within CpG dinucleotides. Lower conversion rates at these sites may indicate actual DNA methylation rather than incomplete bisulfite conversion.

```{r}
rrbs_final <- rrbs_summary_metrics_df |>
  dplyr::select(Sample, PCT_NON_CPG_BASES_CONVERTED, PCT_CPG_BASES_CONVERTED) |>
  tidyr::gather("Site_type", "Percentage", -Sample) |>
  dplyr::mutate(Percentage = ifelse(Percentage <= 1, Percentage, 2-Percentage))

rrbs_width=get_bar_width_with_legend(nrow(rrbs_final$Sample))
rrbs_png=paste0(task_name, ".rrbs_rate.png")

g<-ggplot(rrbs_final, aes(x=Sample, y=Percentage, fill = Site_type)) +
  geom_bar(stat="identity", position=position_dodge()) +
  ylab("BS rate") +
  scale_y_continuous(labels = scales::percent) +  
  theme_classic() +
  mytheme(legend.position = "top")
ggsave(rrbs_png, g, width=rrbs_width, height=4, units="in", dpi=300, bg="white")
```

```{r fig.cap=figRef("bs_cpg_summary", "The BS convert rates of Non-CpG and CpG sites")}
include_graphics(rrbs_png)
```

# Save the session information

```{r include=TRUE, echo=FALSE}
print_table(get_packages_table(), byDT=TRUE, row.names=F)
```
