```{r eval=FALSE}
tau_file="/data/cqs/references/tissue_specific/tau_tissue_specificity_gtex.csv"
gtex_file="/data/cqs/references/protein_atlas/v24.0/rna_tissue_gtex.tsv.zip"
tabula_file = "/data/cqs/references/Tabula_Sapiens/cellxgene_10x3.rds"
```

```{r}
if(!exists("perform_tau")) perform_tau=TRUE
if(!exists("perform_gtex")) perform_gtex=TRUE
if(!exists("perform_tabula")) perform_tabula=TRUE

if(perform_tau) stopifnot(exists("tau_file"))
if(perform_gtex) stopifnot(exists("gtex_file"))
if(perform_tabula) stopifnot(exists("tabula_file"))

stopifnot(exists("all_genes"))
stopifnot(exists("all_genes_label"))
all_genes_fn=gsub('[/:()?\ ]+', "_", all_genes_label)

dataRankTest=function(selectedGenes,alldata,universGenes=NULL,nRep=1000,returnExpRank=FALSE,missingValue=0) {
  if (is.null(universGenes)) {
    universGenes=row.names(alldata)
  }
  expdata=alldata[intersect(row.names(alldata), universGenes),]
  selectedGenes=intersect(selectedGenes,row.names(expdata))
  
  print(paste0("Number of selected genes: ",length(selectedGenes)))
  print(paste0("Number of univers/experiment genes: ",nrow(expdata)))
  
  expdata_norm<-abs(expdata)/(apply(expdata,1,function(x){sqrt(sum(x^2))}))
  expdata_rank<-apply(expdata_norm,2,rank)
  
  topn_rank<-expdata_rank[rownames(expdata_rank) %in% selectedGenes,]
  topn_rank_sum<-apply(topn_rank,2,sum)
  
  if (returnExpRank) {
    return(list(expdata_rank,topn_rank))
  }
  
  perm_rank<-NULL
  for (i in 1:nRep) {
    perm_rank<-rbind(perm_rank,apply(expdata_rank[sample(1:nrow(expdata_rank),nrow(topn_rank)),],2,sum))
  }
  
  pval<-NULL
  
  for (i  in 1:ncol(topn_rank)) {
    pval<-c(pval,sum(perm_rank[,i]>topn_rank_sum[i]))
  }
  
  pval<-pval/nRep
  names(pval)<-colnames(topn_rank)
  
  #return(pval)
  
  expdataCount=apply(expdata,2,function(x) length(which(x!=missingValue)))
  selectedGenesCount=apply(expdata[selectedGenes,],2,function(x) length(which(x!=missingValue)))
  
  return(data.frame(Total=expdataCount,Selected=selectedGenesCount,p=pval,pAdj=p.adjust(pval,method="BH")))
}
```

```{r eval=perform_tau}
cat("\n# Read tau file\n\n")
cat("Tau file was shared by Dr. Kendall Jensen team.\n\n")

tau_df=fread(tau_file, data.table=FALSE)

tau_df$gene_biotype=factor(tau_df$gene_biotype, levels=c("protein_coding", "long_noncoding", "pseudogene", "short_noncoding", "TEC"))

tau_df=tau_df |> dplyr::arrange(gene_name, gene_biotype)

tau_df=tau_df[!duplicated(tau_df$gene_name),] |>
  dplyr::select(-gene_id, -gene_type, -gene_biotype ) |>
  tibble::remove_rownames() |>
  tibble::column_to_rownames("gene_name")

print_table(head(tau_df))

cat(paste0("## Overlap between tau genes and ", all_genes_label, "genes.\n\n"))

genes = list(rownames(tau_df), unique(all_genes))
names(genes)=c("Tau", all_genes_label)

g=ggvenn(genes)

venn_png = paste0("tau_", all_genes_fn, "genes.venn.png")
ggsave(venn_png, g, width=venn_width, height=venn_height, units="in", dpi=300, bg="white")
include_graphics(venn_png)
```

```{r, eval=perform_gtex}
cat("\n# Read protein atlas GTEx gene expression file\n\n")
cat("\nThe file was downloaded from https://www.proteinatlas.org/download/tsv/rna_tissue_gtex.tsv.zip. (v24.0)\n\n")

gtex=fread(gtex_file)
colnames(gtex)[2]="Gene_name"
print_table(head(gtex), row.names=FALSE)

gtex_mat=reshape2::acast(gtex, Gene_name~Tissue, fun.aggregate=sum, value.var="nTPM")
gtex_mat=gtex_mat[rowSums(gtex_mat)>0, ]

cat("\n## Overlap between GTEx genes and ", all_genes_label, " genes\n\n")

genes = list(rownames(gtex_mat), unique(all_genes))
names(genes)=c("GTEx", all_genes_label)

g=ggvenn(genes)

venn_png = paste0("gtex_", all_genes_fn, ".genes.venn.png")
ggsave(venn_png, g, width=venn_width, height=venn_height, units="in", dpi=300, bg="white")
include_graphics(venn_png)
```

```{r eval=perform_tabula}
cat("\n# Load Tabula Sapiens snRNA dataset\n\n")
cat("The tabula sapiens single cell transcriptomic data was downloaded from [cellxgene](https://datasets.cellxgene.cziscience.com/981bcf57-30cb-4a85-b905-e04373432fef.rds).\n\n")
cat("Only the data from 10x 3' platform was used in this report.\n\n")
cat("Read tabula sapiens data from ",basename(tabula_file), ".\n\n")

if(!exists("tabula_obj")){
  tabula_obj=readRDS(tabula_file)
}

cat("\n## Overlap between Tabula Sapiens genes and ", all_genes_label, " genes\n\n")

genes = list(rownames(tabula_obj), unique(all_genes))
names(genes)=c("Tabula snRNA", all_genes_label)

g=ggvenn(genes)

venn_png = paste0("tabula_", all_genes_fn, ".genes.venn.png")
ggsave(venn_png, g, width=venn_width, height=venn_height, units="in", dpi=300, bg="white")
include_graphics(venn_png)

cat("\n## Cell type UMAP\n\n")
cat("Only the cell types with at least 1% of total cells were shown in the UMAP plot.\n\n")

perc_1_cell=nrow(tabula_obj@meta.data) * 0.01
ct=as.data.frame(table(tabula_obj$cell_type)) |>
  dplyr::filter(Freq >= perc_1_cell) |>
  dplyr::arrange(desc(Freq))
ct_obj=tabula_obj[, tabula_obj$cell_type %in% ct$Var1]

g=get_dim_plot_labelby(ct_obj, label.by="cell_type") + theme(plot.title=element_blank())
tabula_celltype_umap_png="tabula_celltype_umap.png"
ggsave(tabula_celltype_umap_png, g, width=14, height=6, units="in", dpi=300, bg="white")
include_graphics(tabula_celltype_umap_png)
```