---
title: "scRNA Sequencing Data Analysis Report"
date: "`r format(Sys.time())`"
output: 
  rmdformats::readthedown:
    code_folding: hide
    self_contained: true
    thumbnails: false
    lightbox: false    
    toc_depth: 3
    number_sections: true
---

<style type="text/css">
#content {
    max-width: 1500px !important;
    margin-left: 300px !important;
}
#table-of-contents {
    width: 300px !important;
}
.main-container {
  max-width: 90%;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 120),
  message = FALSE,
  warning = FALSE,
  echo = TRUE
)

add_cluster<-function(object, new.cluster.name, new.cluster.ids){
  seurat_clusters<-object[["seurat_clusters"]]$seurat_clusters
  names(new.cluster.ids) <- levels(seurat_clusters)

  new.cluster.values<-plyr::mapvalues(x = seurat_clusters, from = levels(seurat_clusters), to = new.cluster.ids)
  names(new.cluster.values)<-names(seurat_clusters)
  
  object[[new.cluster.name]]<-new.cluster.values
  object
}

run_cluster<-function(object, Remove_Mt_rRNA, rRNApattern, Mtpattern, pca_dims, by_sctransform){
  if (by_sctransform) {
    object <- RunPCA(object = object, verbose=FALSE)
  }else{
    if (Remove_Mt_rRNA) {
      rRNA.genes <- grep(pattern = rRNApattern,  rownames(object), value = TRUE)
      Mt.genes<- grep (pattern= Mtpattern,rownames(object), value=TRUE )
      var.genes <- dplyr::setdiff(VariableFeatures(object), c(rRNA.genes,Mt.genes))
    } else {
      var.genes <- VariableFeatures(object)
    }
    object <- RunPCA(object = object, features = var.genes, verbose=FALSE)
  }
  object <- RunUMAP(object = object, dims=pca_dims, verbose = FALSE)
  object <- FindNeighbors(object = object, dims=pca_dims, verbose=FALSE)
  object <- FindClusters(object=object, verbose=FALSE, random.seed=random.seed, resolution=resolution)

  if (by_sctransform) {
    markers <- FindAllMarkers(object, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
  }else{
    markers <- FindAllMarkers(object, features=var.genes, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
  }
  markers <- markers[markers$p_val_adj < 0.01,]
  return(list(object=object, markers=markers))
}

ORA_celltype<-function(medianexp,cellType,weight){
  ORA_result<-matrix(NA, nrow=length(cellType),ncol=dim(medianexp)[2])
  CTA_result<-matrix(0,nrow=length(cellType),ncol=dim(medianexp)[2])
  exp_z<-scale(medianexp)
  genenames<-rownames(medianexp)   
  for (j in 1: dim(medianexp)[2]){
    clusterexp<-medianexp[,j] 
    clusterexp_z<-exp_z[,j]
    for (i in 1:length(cellType)){
      
      ct_exp<-length(intersect(genenames[clusterexp>0],cellType[[i]]))
      ct_not_exp<-length(cellType[[i]])-ct_exp
      exp_not_ct<-sum(clusterexp>0)-ct_exp
      not_exp_not_ct<-length(clusterexp)-ct_not_exp 
      cont.table<-matrix(c(ct_exp,ct_not_exp,exp_not_ct,not_exp_not_ct),nrow=2)
      ORA_result[i,j]<-fisher.test(cont.table,alternative="greater")$p.value
      ###
      weight_ss<-weight[names(weight)%in%cellType[[i]]]
      ind<-match(names(weight_ss),genenames)
      exp_ss<-clusterexp_z[ind[!is.na(ind)]]
      weight_ss<-weight_ss[!is.na(ind)]
      CTA_result[i,j]<-sum(exp_ss*weight_ss)/(length(exp_ss)^(1/3))
    }
  }
  rownames(ORA_result)<-rownames(CTA_result)<-names(cellType)
  minp_ora_ind<- apply(ORA_result,2,function(x){which.min(x)})
  minp_ora<-apply(ORA_result,2,min)
  names(minp_ora)<-rownames(ORA_result)[minp_ora_ind]
  
  max_cta_ind<- apply(CTA_result,2,function(x){which.max(x)})
  max_cta<-apply(CTA_result,2,max,na.rm=T)
  names(max_cta)<-rownames(CTA_result)[max_cta_ind]
  return(list(ora=ORA_result,cta=CTA_result,min_ora=minp_ora,max_cta=max_cta))
}
```

```{r, child="Functions.Rmd"} 
```

```{r init}
library(dplyr)
library(Seurat)
library(knitr)
library(kableExtra)
library(ggplot2)
library(ggpubr)
library(rmdformats)
library(DT)
library(data.table)
library(digest)
library(heatmap3)
library(cowplot)
library(scales)
library(stringr)

options(future.globals.maxSize= 10779361280)
random.seed=20200107

options_table<-read.table("fileList2.txt", sep="\t", header=F, stringsAsFactors = F)
options<-split(options_table$V1, options_table$V2)

dynamic_options<-options_table[!(options_table$V2 %in% c("details_rmd")),]
optionstr<-paste0(dynamic_options$V1, collapse = "_")

Mtpattern= options$Mtpattern
rRNApattern=options$rRNApattern
Remove_Mt_rRNA= ifelse(options$Remove_Mt_rRNA == "FALSE", FALSE, TRUE)
resolution=as.numeric(options$resolution)
species=options$species
markerfile<-options$markers_file
details_rmd<-options$details_rmd

by_integration<-ifelse(options$by_integration == "0", FALSE, TRUE)
by_sctransform<-ifelse(options$by_sctransform == "0", FALSE, TRUE)

prefix<-options$prefix

hasDetails<-file.exists(details_rmd)
if(hasDetails){
  hasDetails<-file.copy(details_rmd, "Details.Rmd", overwrite=T)
}

nFeature_cutoff_min=as.numeric(options$nFeature_cutoff_min)
nFeature_cutoff_max=as.numeric(options$nFeature_cutoff_max)
nCount_cutoff=as.numeric(options$nCount_cutoff)
mt_cutoff=as.numeric(options$mt_cutoff)
#nCount_sd_cutoff=as.numeric(options$nCount_sd_cutoff)
nCount_sd_cutoff=0

pca_dims<-1:as.numeric(options$pca_dims)

filelist1<-read.table("filelist1.txt", header=F)
filestr<-paste0(filelist1$V1, collapse = "_")

prefixMd5=digest(paste0(optionstr, filestr), "md5", serialize = FALSE)

finalList<-list()
finalListFile<-paste0(prefix, ".final.rds")
finalListFileOptionsMd5<-paste0(prefix, ".final.options.md5")

isCalc=TRUE
if(file.exists(finalListFile) && file.exists(finalListFileOptionsMd5)){
  oldMd5=str_trim(readChar(finalListFileOptionsMd5, file.info(finalListFileOptionsMd5)$size))
  isCalc=(oldMd5 != prefixMd5)
}

toMouseGeneSymbol<-function(x){
  result=paste0(toupper(substr(x,1,1)),tolower(substr(x,2,nchar(x))))
  return(result)
}

if(isCalc){
  #preparing cell activity database
  marker<-data.frame(fread(markerfile))
  hsind<-regexpr(species,marker[,1])
  marker_species<-marker[hsind>0 & marker$ubiquitousness.index<0.05,]
  if (species=="Mm") {
    ##change the gene symbol only keep the first letter capitalize
    marker_species$official.gene.symbol<-toMouseGeneSymbol(marker_species$official.gene.symbol)
  }
  if (species=="Hs") {
    marker_species$official.gene.symbol<-toupper(marker_species$official.gene.symbol)
  }
  cellType<-tapply(marker_species$official.gene.symbol,marker_species$cell.type,list)
  freq<-sort((table(unlist(cellType)))/length(cellType))
  weight<-1+sqrt((max(freq)-freq)/(max(freq)-min(freq)))
  finalList$cell_activity_database<-list(cellType=cellType, weight=weight)
    
  #read raw count dat
  filelist1<-read.table("filelist1.txt", header=F, stringsAsFactors = F)
  rawobjs<-apply(filelist1, 1, function(x){
    fileName  = x[[1]]
    fileTitle = x[[2]]
    if(dir.exists(fileName)){
      counts = Read10X(fileName)
    } else {
      counts = Read10X_h5(fileName)
    }
    
    if (is.list(counts)){
      counts<-counts$`Gene Expression` 
    }

    if (species=="Mm") {
      rownames(counts)<-toMouseGeneSymbol(rownames(counts))
    }
    if (species=="Hs") {
      rownames(counts)<-toupper(rownames(counts))
    }
    sobj = CreateSeuratObject(counts = counts, project = fileTitle)
    sobj[["percent.mt"]] <- PercentageFeatureSet(object = sobj, pattern = Mtpattern)
    return(sobj)
  })
  if(length(rawobjs) == 1){
    rawobj <- rawobjs[[1]]
  }else{
    cellids<-unlist(lapply(rawobjs, function(x) x@project.name))
    rawobj <- merge(rawobjs[[1]], y = unlist(rawobjs[2:length(rawobjs)]), add.cell.ids = cellids, project = "integrated")
  }
  
  finalList$rawobj<-rawobj

  #filter cells
  finalList$filter<-list(nFeature_cutoff_min=nFeature_cutoff_min,
                        nFeature_cutoff_max=nFeature_cutoff_max,
                        mt_cutoff=mt_cutoff,
                        nCount_cutoff=nCount_cutoff,
                        nCount_sd_cutoff=nCount_sd_cutoff)
  objs<-lapply(rawobjs, function(x){
    sobj<-subset(x, subset = nFeature_RNA > nFeature_cutoff_min & nFeature_RNA<nFeature_cutoff_max & nCount_RNA > nCount_cutoff & percent.mt < mt_cutoff)

    if(nCount_sd_cutoff > 0){
      nCount_mean = mean(sobj[["nCount_RNA"]])
      nCount_sd = sd(sobj[["nCount_RNA"]])
      nCount_sd_min = nCount_mean - nCount_sd_cutoff * nCount_sd
      nCount_sd_max = nCount_mean + nCount_sd_cutoff * nCount_sd
      sobj<-subset(sobj, subset = nCount_RNA > nCount_sd_min & nCount_RNA < nCount_sd_max)
      finalList$filter$nCount_sd_min = nCount_sd_min
      finalList$filter$nCount_sd_max = nCount_sd_max
    }

    sobj<-RenameCells(object=sobj, add.cell.id = unique(sobj$orig.ident))
    return(sobj)
  })  
  rm(rawobjs)

  if(by_integration){
    if(by_sctransform){
      #perform sctransform
      objs<-lapply(objs, function(x){
        x <- SCTransform(x, verbose = FALSE)
        return(x)
      })  
      
      if(length(objs) == 1){
        obj <- objs[[1]]
      }else{
        obj_features <- SelectIntegrationFeatures(object.list = objs, nfeatures = 3000)
        objs <- PrepSCTIntegration(object.list = objs, anchor.features = obj_features, verbose = FALSE)
        obj_anchors <- FindIntegrationAnchors(object.list = objs, normalization.method = "SCT", anchor.features = obj_features, verbose = FALSE)
        obj <- IntegrateData(anchorset = obj_anchors, normalization.method = "SCT", verbose = FALSE)      
      }
    }else{
      #perform standard workflow
      objs<-lapply(objs, function(x){
        x <- NormalizeData(x, verbose = FALSE)
        x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000, verbose = FALSE)  
        return(x)
      })  
      
      if(length(objs) == 1){
        obj <- objs[[1]]
      }else{
        obj.anchors <- FindIntegrationAnchors(object.list = objs, dims = 1:20)
        obj <- IntegrateData(anchorset = obj.anchors, dims = 1:20)    
        obj <- ScaleData(obj, verbose = FALSE)
      }
    }
  }else{
    if(length(objs) == 1){
      obj <- objs[[1]]
    }else{
      obj <- merge(objs[[1]], y = unlist(objs[2:length(objs)]), project = "integrated")
    }
    if(by_sctransform){
      #perform sctransform
      obj <- SCTransform(obj, verbose = FALSE)
    }else{
      obj <- NormalizeData(obj, verbose = FALSE)
      obj <- FindVariableFeatures(obj, selection.method = "vst", nfeatures = 2000, verbose = FALSE)
      all.genes <- rownames(obj)  
      obj <- ScaleData(obj, features = all.genes, verbose = FALSE)
    }
  }
  
  obj_markers <- run_cluster(obj, Remove_Mt_rRNA, rRNApattern, Mtpattern, pca_dims, by_sctransform)
  
  finalList$markers<-obj_markers$markers
  obj<-obj_markers$object
  seurat_clusters<-unlist(obj[["seurat_clusters"]])
  seurat_colors<-hue_pal()(length(levels(seurat_clusters)))
  names(seurat_colors)<-levels(seurat_clusters)
  
  finalList$seurat_colors<-seurat_colors
  
  clusters<-obj@active.ident
  sumcounts<-t(apply(GetAssayData(obj,assay="RNA",slot="counts"),1,function(x){tapply(x,clusters,sum)}))
  logsumcounts<-log2(sumcounts+1)
  data.quantileAll <- apply(logsumcounts, 2, function(x){quantile(x, 0.75)})
  
  norm_method=""
  if(any(data.quantileAll == 0)){
    norm_method = ".normByTotal"
    data.all <- apply(logsumcounts, 2, sum)
    data.all<-data.all / median(data.all)
    data.norm <- t(t(logsumcounts) / data.all)
  }else{
    norm_method = ".normByUpQuantile"
    data.quantileAll<-data.quantileAll / median(data.quantileAll)
    data.norm <- t(t(logsumcounts) / data.quantileAll)
  }

  colnames(sumcounts)<-paste0("Cluster", colnames(sumcounts))
  write.csv(sumcounts, file=paste0(prefix, ".cluster.count.csv"))
  
  oldname<-colnames(data.norm)
  colnames(data.norm)<-paste0("Cluster", oldname)
  write.csv(data.norm, file=paste0(prefix, ".cluster", norm_method, ".csv"))
  colnames(data.norm)<-oldname
  
  predict_celltype<-ORA_celltype(data.norm,finalList$cell_activity_database$cellType,finalList$cell_activity_database$weight)
  
  new.cluster.ids<-names(predict_celltype$max_cta)
  names(new.cluster.ids) <- levels(seurat_clusters)
  obj[["cellactivity_clusters"]] <- new.cluster.ids[unlist(obj[["seurat_clusters"]])]
  
  clusterDf<-data.frame(seurat=unlist(obj[["seurat_clusters"]]), cellactivity=unlist(obj[["cellactivity_clusters"]]))
  clusterDf$seurat_colors<-seurat_colors[clusterDf$seurat]
  
  clusterDf$seurat_cellactivity<-paste0(clusterDf$seurat, " : ", clusterDf$cellactivity)
  seurat_cellactivity<-clusterDf$seurat_cellactivity
  
  caCount<-table(clusterDf$cellactivity)
  clusterDf$caCount<-caCount[clusterDf$cellactivity]
  
  clusterDf<-clusterDf[order(-clusterDf$caCount, clusterDf$seurat),]
  seurat_cellactivity<-factor(seurat_cellactivity, levels=unique(clusterDf$seurat_cellactivity))
  seurat_cellactivity_colors<-unique(clusterDf$seurat_colors)
  obj[["seurat_cellactivity_clusters"]] <- seurat_cellactivity
  
  clusters<-data.frame("cell" = c(1:length(obj$seurat_clusters)), "seurat_clusters"=as.numeric(as.character(obj$seurat_clusters)), "cellactivity_clusters"=obj$cellactivity_clusters, "seurat_cellactivity_clusters"=obj$seurat_cellactivity_clusters, stringsAsFactors = F)
  rownames(clusters)<-names(obj$seurat_clusters)
  write.csv(clusters, file=paste0(prefix, ".cluster.csv"))
    
  cta_index<-apply(predict_celltype$cta,2,function(x){return(order(x,decreasing=T)[1:2])})
  cta_index<-unique(sort(cta_index))
  
  cta_mat<- predict_celltype$cta[cta_index,]
  colnames(cta_mat)<-paste0(names(predict_celltype$max_cta), " : ", c(0:(ncol(cta_mat)-1)))
  
  ora_mat<- predict_celltype$ora[cta_index,]
  ora_mat<--log10(ora_mat)  
  colnames(ora_mat)<-colnames(cta_mat)
  
  finalList$cell_activity_database$predicted<-predict_celltype
  finalList$seurat_cellactivity_colors<-seurat_cellactivity_colors
  finalList$cta_mat<-cta_mat
  finalList$ora_mat<-ora_mat
  finalList$obj<-obj
  saveRDS(finalList, file=finalListFile)
  
  fileConn<-file(finalListFileOptionsMd5)
  writeLines(c(prefixMd5), fileConn)
  close(fileConn)
}else{
  finalList<-readRDS(file=finalListFile)
}
obj<-finalList$obj
seurat_colors<-finalList$seurat_colors
seurat_cellactivity_colors<-finalList$seurat_cellactivity_colors
hasMultipleSamples<-length(unique(obj@meta.data$orig.ident))>1
```

# Quality control

A few QC metrics [commonly used](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4758103/) by the community include

* The number of unique genes detected in each cell. 
    + Low-quality cells or empty droplets will often have very few genes
    + Cell doublets or multiplets may exhibit an aberrantly high gene count
* Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)
* The percentage of reads that map to the mitochondrial genome
    + Low-quality / dying cells often exhibit extensive mitochondrial contamination
    + We calculate mitochondrial QC metrics with the `PercentageFeatureSet` function, which calculates the percentage of counts originating from a set of features
    + We use the set of all genes starting with either `mt-` or `MT-` as a set of mitochondrial genes

We visualize QC metrics.

```{r qc2, fig.height=6, fig.width=13, fig.cap=figRef("qc_metrics_2", "Scatter plot of number of feature/percentage of mitochondrial vs number of read count")}
plot1 <- FeatureScatter(object = finalList$rawobj, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_hline(yintercept = mt_cutoff, color="black")  + geom_vline(xintercept = nCount_cutoff, color="black")
plot2 <- FeatureScatter(object = finalList$rawobj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")  + geom_hline(yintercept = c( nFeature_cutoff_min, nFeature_cutoff_max), color="black")  + geom_vline(xintercept = nCount_cutoff, color="black") 
CombinePlots(plots = list(plot1,plot2))
```

# Data filter and cluster analysis 

We use QC metrics to filter cells.

* We filter cells that have unique feature counts less than `r nFeature_cutoff_min` or larger than `r nFeature_cutoff_max` 
* We filter cells that have read count less than `r nCount_cutoff`
* We filter cells that have >`r mt_cutoff`% mitochondrial counts

```{r data_filter, results="asis"}
rawCells<-data.frame(table(finalList$rawobj$orig.ident))
filteredCells<-data.frame(table(finalList$obj$orig.ident))
qcsummary<-merge(rawCells, filteredCells, by = "Var1")
colnames(qcsummary)<-c("Sample", "RawCell", "ValidCell")
qcsummary$DiscardCell<-qcsummary$RawCell-qcsummary$ValidCell
qcsummary$DiscardRate<-qcsummary$DiscardCell / qcsummary$RawCell
print(kable_styling(kable(qcsummary, caption=tabRef("qcsummary", "Cell count in sample"))))
```

## Marker genes for each cluster

```{r markers, results="asis"}
clusterMarkers<-finalList$markers %>% group_by(cluster)
write.csv(clusterMarkers, file=paste0(prefix, ".allmarkers.csv"), row.names=F, quote = F)

top10 <- finalList$markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
top10marker_file = paste0(prefix, ".top10markers.csv")
write.csv(top10, file=top10marker_file, row.names=F, quote = F)

top5 <- finalList$markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_logFC)
print(kable_styling(kable(top5, caption=tabRef("top5", "Top 5 marker genes in each cluster"))))
```

# Cell clusters
    
```{r qc1, fig.height=5, fig.width=13, fig.cap=figRef("qc_metrics_1", "Violin plot of percentage of mitochondrial, number of feature and number of read count")}
g<-VlnPlot(object = obj, features = c("percent.mt", "nFeature_RNA", "nCount_RNA"))
print(g)
```

```{r umap_cells, fig.height=8, fig.width=18, fig.cap=figRef("umap_cells", ifelse(hasMultipleSamples, "UMAP of cells (left) and samples (right)", "UMAP of cells"))}
# note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters
p1<-DimPlot(object = obj, reduction = 'umap', label=TRUE, group.by="seurat_clusters", cols=seurat_colors)
if(hasMultipleSamples){
  p2<-DimPlot(object = obj, reduction = 'umap', label=FALSE, group.by="orig.ident")
  plot_grid(p1,p2)
}else{
  print(p1)
}
```

```{r sample_cells_count, eval=hasMultipleSamples, fig.height=8, fig.width=18, fig.cap=figRef("sample_cells_count", "Percentage of cell from samples in each cluster (total cell in each sample were normalized to equal number)")}
os<-table(obj$orig.ident, obj$seurat_clusters)
osn<-os / rowSums(os) * 5000
osm<-melt(osn)
colnames(osm)<-c("Sample", "Cluster", "Cell")
osm$Cluster<-paste0("Cluster ", osm$Cluster)
osm$Cluster<-factor(osm$Cluster, levels=unique(osm$Cluster))
osm_perc = osm %>% dplyr::group_by(Cluster) %>% dplyr::mutate(Percent = Cell/sum(Cell))
ggplot(osm_perc, aes(x="",y=Percent,fill=Sample)) + geom_bar(stat="identity", width=1) + coord_polar("y", start=0) + facet_wrap(~Cluster) + theme_void()

```

```{r heatmap_samples, fig.height=max(10, length(top10$gene) / 10), fig.width=12, fig.cap=figRef("heatmap_samples", "Marker genes in each cell type")}
DoHeatmap(obj, features = top10$gene, group.colors=seurat_colors, angle = 90) + NoLegend()
```

# Predict cell types

## Cell type prediction using gene activity database

### Predicted cell types for each cluster by cell activity analysis

```{r gene_activity_tbl, results="asis"}
predicted<-finalList$cell_activity_database$predicted
cta_table<-data.frame(Cluster=c(1:length(predicted$max_cta))-1, 
                      CellType=names(predicted$max_cta),
                      CtaScore=round(predicted$max_cta * 10) / 10.0,
                      stringsAsFactors = F)
cta_table$OraPvalue=apply(cta_table, 1, function(x){
  ct=x[2]
  cl=as.numeric(x[1]) + 1
  predicted$ora[ct, cl]
})

cluster_sample<-as.data.frame.matrix(table(obj$seurat_clusters, obj$orig.ident))

nc<-apply(cluster_sample, 2, function(x){
  tc=sum(x)
  perc<-x/tc
  return(round(perc*1000) / 10.0)
})
colnames(nc)<-paste0(colnames(nc), "_perc")

cta_combined<-cbind(cta_table, cluster_sample, nc)
write.csv(cta_combined, file=paste0(prefix, ".cluster.summary.csv"), row.names=F, quote=F)

print(kable_styling(kable(cta_combined, caption=tabRef("cta_table", "Cell type prediction result"))))
```

```{r gene_activity_umap, fig.width=18, fig.height=10, results="asis", fig.cap=figRef("umap_cluster_cell", "Predicted cell types")}
p2<-DimPlot(object = obj, reduction = "umap",label=T, group.by="seurat_cellactivity_clusters", cols=seurat_cellactivity_colors)
print(p2)
```

```{r gene_activity_umap_samples, eval=hasMultipleSamples, fig.height=8, fig.width=18, fig.cap=figRef("gene_activity_umap_samples", "Predicted cell types by sample")}
DimPlot(obj, reduction = "umap", label=T, group.by="seurat_clusters", cols=seurat_colors, split.by = "orig.ident") + NoLegend()
```

```{r gene_activity_heatmap, fig.width=18, fig.height=max(10, length(top10$gene) / 10), results="asis", fig.cap=figRef("cell_activity_heatmap", "Predicted cell type cluster heatmap")}
DoHeatmap(obj, features = top10$gene, group.by="seurat_cellactivity_clusters", group.colors=seurat_cellactivity_colors, angle=90) + NoLegend()
```

```{r gene_activity_cta_score_heatmap, fig.width=18, fig.height=10, results="asis", fig.cap=figRef("cell_activity_cta_heatmap", "Predicted cell type activity score")}
heatmap3(finalList$cta_mat, scale="none", margins=c(10,5))
```

```{r, child="Details.Rmd", eval=hasDetails} 
```

# Project options

```{r options, echo=F}
if(hasDetails){
  file.remove("Details.Rmd")
}
colnames(options_table)<-c("Value", "Option")
options_table<-options_table[,c("Option", "Value")]
options_table<-options_table[!(options_table$Option %in% c("Mtpattern", "rRNApattern")),]
kable(options_table)
```
