---
title: "scRNA Sequencing Data Analysis Report"
date: "`r format(Sys.time())`"
output: 
  rmdformats::readthedown:
    code_folding: hide
    self_contained: true
    thumbnails: false
    lightbox: false    
    toc_depth: 3
    number_sections: true
---

<style type="text/css">
#content {
    max-width: 1500px !important;
    margin-left: 300px !important;
}
#table-of-contents {
    width: 300px !important;
}
.main-container {
  max-width: 90%;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 120),
  message = FALSE,
  warning = FALSE,
  echo = FALSE
)

source("scRNA_func.r")

```

```{r, child="Functions.Rmd", include=FALSE} 
```

```{r init, include=FALSE}
library(dplyr)
library(Seurat)
library(knitr)
library(kableExtra)
library(ggplot2)
library(ggpubr)
library(rmdformats)
library(DT)
library(data.table)
library(digest)
library(heatmap3)
library(cowplot)
library(scales)
library(stringr)
library(htmltools)
library(patchwork)
require(data.table)

options(future.globals.maxSize= 10779361280)

options_table<-read.table("fileList2.txt", sep="\t", header=F, stringsAsFactors = F)
myoptions<-split(options_table$V1, options_table$V2)

has_summary_layer<-!is.null(myoptions$summary_layer_file)
has_rename_file<-!is.null(myoptions$rename_png)

prefix<-myoptions$prefix

filelist1<-read.table("fileList1.txt", header=F)
files<-split(filelist1$V1, filelist1$V2)

celltype<-read.csv(file=files$activity_celltype)
celltype_db<-readRDS(file=files$activity_rds)

has_chetah<-"chetah_png" %in% names(files)
has_signac<-"signac_png" %in% names(files)

sample_cell<-read.csv(files$sample_cell_csv)

hasMultipleSamples<-nrow(sample_cell)>1

samples<-sample_cell$Sample
if(length(samples) < 3){
  sampleHeight=4
}else{
  sampleHeight=min(10, ceiling(length(samples) / 3) * 4)
}
sampleCol=ceiling(sqrt(length(samples)))
sampleRow=ceiling(length(samples) / sampleCol)

isLargeDataset=sum(sample_cell$ValidCell) > 30000
if(isLargeDataset){
  downsample=round(300/nrow(celltype)) * 100
}

has_hto = file.exists("fileList4.txt")

hasBiomarker<-file.exists(files$celltype_markers_csv)
annotate_tcell<-ifelse(myoptions$annotate_tcell=='0', FALSE, TRUE)
```

```{r hto, results="asis", eval=has_hto, fig.width=12, fig.height=10}
cat("# HTO Demultiplex \n\n")
hto_files = read.table("fileList4.txt", sep="\t")
file_map = split(hto_files$V1, hto_files$V2)
for(sample in sort(names(file_map))){
  cat("## ", sample, "\n\n")
  hto_file=file_map[[sample]]
  cat("![](",hto_file,")\n\n")
}
```

# Quality control

A few QC metrics [commonly used](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4758103/) by the community include

* The number of unique genes detected in each cell. 
    + Low-quality cells or empty droplets will often have very few genes
    + Cell doublets or multiplets may exhibit an aberrantly high gene count
* Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)
* The percentage of reads that map to the mitochondrial genome
    + Low-quality / dying cells often exhibit extensive mitochondrial contamination
    + We calculate mitochondrial QC metrics with the `PercentageFeatureSet` function, which calculates the percentage of counts originating from a set of features
    + We use the set of all genes starting with either `mt-` or `MT-` as a set of mitochondrial genes

We visualize QC metrics.

```{r qc1, fig.height=6, fig.width=13, fig.cap=figRef("qc_metrics_2", "Scatter plot of number of feature/percentage of mitochondrial vs number of read count")}
include_graphics(files$seurat_qc_1_png)
```

```{r qc2, fig.height=sampleHeight, fig.width=13, fig.cap=figRef("qc_metrics_2_1", "Density plot of number of read vs percentage of mitochondrial")}
include_graphics(files$seurat_qc_2_png)
```

```{r qc3, fig.height=sampleHeight, fig.width=13, fig.cap=figRef("qc_metrics_2_2", "Density plot of number of feature vs percentage of mitochondrial")}
include_graphics(files$seurat_qc_3_png)
```

# Data preprocessing and integration 

We use QC metrics to filter cells.

* We filter cells that have unique feature counts less than `r myoptions$nFeature_cutoff_min` or larger than `r myoptions$nFeature_cutoff_max` 
* We filter cells that have read count less than `r myoptions$nCount_cutoff`
* We filter cells that have >`r myoptions$mt_cutoff`% mitochondrial counts

## Cell count in sample

```{r data_filter, results="asis"}
print(kable(sample_cell, caption=tabRef("qcsummary", "Cell count in sample")) %>% kable_styling() %>% htmltools::HTML())

```

## QC after preprocessing

```{r qc4, fig.height=5, fig.width=13, fig.cap=figRef("qc_metrics_1", "Violin plot of percentage of mitochondrial, number of feature and number of read count")}
include_graphics(files$seurat_qc_4_png)
```

## UMAP after sample integration

```{r merge_umap, fig.width=13, fig.cap=figRef("merge_umap", "UMAP after sample integration")}
include_graphics(files$seurat_merge_png)
```

```{r markers, results="asis", eval=hasBiomarker}
max_markers<-read.csv(files$celltype_markers_csv)
max_markers<-max_markers[order(max_markers$cluster),]

top5 <- max_markers %>% group_by(cluster) %>% top_n(n = 5, wt = .data[["avg_log2FC"]])
  
cat("## Marker genes for each cluster\n")
print(kable(top5, caption=tabRef("top5", "Top 5 marker genes in each cluster")) %>% kable_styling() %>% htmltools::HTML())
```


# Cell clusters
    
```{r umap_cells, fig.height=8, fig.width=18, fig.cap=figRef("umap_cells", ifelse(hasMultipleSamples, "UMAP of cells (left) and samples (right)", "UMAP of cells"))}
include_graphics(files$cluster_umap_sample_cell_png)
```

## Percentage of cell in each cluster (total cell in each sample were normalized to equal number)

```{r cluster_sample_percByCluster_png, eval=hasMultipleSamples, fig.height=18, fig.width=18, fig.cap=figRef("cluster_sample_percByCluster_png", "Percentage of cell in each cluster (total cell in each sample were normalized to equal number)")}
include_graphics(files$cluster_sample_percByCluster_png)

```

## Percentage of cell in each sample

```{r cluster_sample_percBySample_png, eval=hasMultipleSamples, fig.height=18, fig.width=18, fig.cap=figRef("cluster_sample_percBySample_png", "Percentage of cell in each sample")}
include_graphics(files$cluster_sample_percBySample_png)

```

# Cell type annotation

## Cell type annotation using gene activity database


```{r gene_activity_cta_score_heatmap, fig.width=12, fig.height=10, results="asis", fig.cap=figRef("cell_activity_cta_heatmap", "Predicted cell type activity score")}
celltype_db<-readRDS(files$activity_rds)
heatmap3(celltype_db$cell_activity_database$cta_mat, scale="none", margins=c(10,5))
```


```{r gene_activity_tcell_cta_score_heatmap, fig.width=12, fig.height=10, results="asis", eval=annotate_tcell, fig.cap=figRef("tcell_activity_cta_heatmap", "Predicted T cell subtype activity score")}
heatmap3(celltype_db$tcell_activity_database$cta_mat, scale="none", margins=c(10,5))
```

```{r gene_activity_tbl, results="asis"}
# 
# cta_combined<-get_cta_combined(obj, celltype_db$cell_activity_database$predicted)
# 
# print(kable(cta_combined, row.names=F, caption=tabRef("cta_table", "Cell type prediction result")) %>% kable_styling() %>% htmltools::HTML())
```

```{r tcell_gene_activity_tbl, results="asis", eval=annotate_tcell}
# cta_combined<-get_cta_combined(obj, celltype_db$tcell_activity_database$predicted)
# 
# write.csv(cta_combined, file=paste0(prefix, ".tcell.cluster.summary.csv"), row.names=F, quote=F)
# 
# print(kable(cta_combined, row.names=F,  caption=tabRef("tcell_cta_table", "T cell subtype prediction result")) %>% kable_styling() %>% htmltools::HTML())
```

```{r gene_activity_umap_layer, fig.width=12, fig.height=8, results="asis", fig.cap=figRef("umap_cluster_cell_layer", "Predicted cell types layer 1"), eval=has_summary_layer}
include_graphics(files$activity_summary_layer_png)
```

```{r gene_activity_umap, fig.width=12, fig.height=8, results="asis", fig.cap=figRef("umap_cluster_cell", "Predicted cell types")}
include_graphics(files$activity_cluster_png)
```

```{r gene_activity_group_umap, fig.width=12, eval=file_not_empty(files$activity_group_png), fig.height=8, results="asis", fig.cap=figRef("umap_cluster_cell_group", "Predicted cell types by groups")}
include_graphics(files$activity_group_png)
```

# Marker genes

## Cell type specific marker genes

```{r gene_activity_bubble_map, fig.width=12, fig.height=8, results="asis", fig.cap=figRef("gene_activity_bubble_map", "Bubble map of marker genes")}
include_graphics(files$bubblemap_png)
```

## Cluster specific marker genes

### Cluster specific marker gene heatmap


```{r eval=isLargeDataset, comment=''}
cat("There are total ", sum(sample_cell$ValidCell), " cells, so each cluster was downsampled to ", downsample, " cells for heatmap.")
```

```{r cluster_marker_heatmap, fig.width=12, fig.height=8, results="asis", eval=file_not_empty(files$celltype_markers_heatmap_png), fig.cap=figRef("cluster_marker_heatmap", "Cluster marker genes heatmap")}
include_graphics(files$celltype_markers_heatmap_png)
```


```{r marker_genes, results="asis", fig.width=12, fig.height=10}
marker_files = read.csv(files$celltype_markers_png_list)

ac<-read.csv(files$activity_celltype)
acmap<-split(ac$cell_type, ac$seurat_clusters)

for(idx in c(1:nrow(marker_files))){
  cluster<-marker_files[idx, "Cluster"]
  marker_file=marker_files[idx, "Dotfile"]
  
  celltype<-unlist(acmap[as.character(cluster)])
  
  cat("### Cluster ", cluster, " : ", celltype, "\n\n")
  cat("![](",marker_file,")\n\n")
}

```


```{r gene_activity_umap_samples, eval=hasMultipleSamples, fig.height=max(6, ceiling(12*sampleRow/sampleCol)), fig.width=12, fig.cap=figRef("gene_activity_umap_samples", "Predicted cell types by sample")}
# g=DimPlot(obj, reduction = "umap", label=T, group.by="seurat_clusters", cols=seurat_colors, split.by = "orig.ident", combine=F)
# g<-g[[1]]
# g2=g+facet_wrap(~orig.ident, ncol=sampleCol) + theme_bw() + theme(strip.background = element_blank()) + NoLegend()
# print(g2)
```


```{r chetah_png, fig.width=18, fig.height=10, results="asis", eval=has_chetah, fig.cap=figRef("chetah_png", "Predicted cell type by CHETAH")}
cat("## CHETAH prediction\n\n")
include_graphics(files$chetah_png)
```


```{r signac_png, fig.width=18, fig.height=10, results="asis", eval=has_signac, fig.cap=figRef("signac_png", "Predicted cell type by SIGNAC")}
cat("## Signac prediction\n\n")
include_graphics(files$signac_png)
```

```{r rename_png, fig.width=18, fig.height=10, results="asis", eval=has_rename_file, fig.cap=figRef("rename_png", "Rename cell type by expert")}
cat("## Renamed cell type\n\n")
include_graphics(files$rename_png)
```

# Project options

```{r options, results="asis"}
colnames(options_table)<-c("Value", "Option")
options_table<-options_table[,c("Option", "Value")]
options_table<-options_table[!(options_table$Option %in% c("Mtpattern", "rRNApattern")),]
print(kable(options_table, caption=tabRef("options_table", "Options table")) %>% kable_styling() %>% htmltools::HTML())
```
