---
title: "singlecell_multiple_analysis"
author: Qi Liu
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_depth: 4
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,  message=FALSE, warning=FALSE, results = 'hide', fig.width=7, fig.height = 7,tidy = TRUE, tidy.opts = list(comment = FALSE))
```





```{r, echo=FALSE, message=FALSE, warning=FALSE}
#load the library and set the parameters and load our own function
library(Seurat)
library(data.table)
library(dplyr)
library(ggplot2)
library(pheatmap)
library(RCurl)



#library(batchelor)
#library(scran)
```





```{r, echo=FALSE}

cat("\n\n## Files Parameters\n\n")

#raw scRNAseq data files and clinical information
#work directory to save or load objectlist, objectintegrated and integratedmarkers; the raw files are in workdir/raw

workdir<-"/Users/qiliu/Dropbox (VUMC)/KaseyVicker/LDLR/"

#raw data file locations
filelist<-list.files(paste0(workdir,"raw"),include.dirs = F,recursive = T,pattern="*.h5$",full.names = T)
species="Mm" # Hs or Mm
transpose=FALSE  #if row is gene, column is cell, then transpose=FALSE; if column is gene, then transpose=TRUE
Ensemblfile=NULL
  #"/Users/qiliu/Dropbox (VUMC)/PCA/gencode.v34.annotation.single_PAR_gene_names_ids.csv" # file used to convert ensembl ID to gene symbol, the first column is gene symbol, the second is ensembl ID; IF NULL, then don't convert


#clinical Info including SampleId and other clincal Information; SampleId is used to link to the filelist, should be character
clinicalInfo<-data.frame(SampleId=c("DKO_F","DKO_M","LDLR_F","LDLR_M"),conditions=c("DKO","DKO","LDLR","LDLR"), gender=c("F","M","F","M"),stringsAsFactors = FALSE)
clinicalfeature<-c(1,1)

#combine them into one sample Info table

SampleInfo<-data.frame(clinicalInfo,countfile=filelist,stringsAsFactors=F)




##SampleId and countfile are required, other experiment info can also be added.

```


```{r, echo=FALSE}

##analysis individual or integrated or generate the final report
analysis_individual = TRUE  # if True, analysis individual, generate analysis report for each individual and save each individual to objectlist.Rdata. 
analysis_integrated = FALSE # if true, load objectlist.Rdata to analysis, generate one combined object objectintegrated.Rdata
batchcorrection =FALSE
Finalreport=!analysis_individual & !analysis_integrated  #if true, generate the final report on the combined objectintegrated.Rdata


```



```{r,echo=FALSE}

##cutoff dataframe for each sample to filter empty droplets
Cutoff<-data.frame(nFeature_cutoff_min = 500 ,nFeature_cutoff_max = 7000,nCount_cutoff=1000, mt_cutoff = 10, cluster_remove=c(""),stringsAsFactors = F)

#every sample share the same cutoff, otherwise put every parameter in the Cutoff dataframe
if (dim(Cutoff)[1]==1)
{Cutoff<-data.frame(lapply(Cutoff, rep, nrow(SampleInfo)),stringsAsFactors = F)}


Mtpattern= "^MT-|^mt-"
rRNApattern="^Rp[sl][[:digit:]]|^RP[SL][[:digit:]]"
Remove_Mt_rRNA=FALSE

resolution=0.3



#genelistfile<-"/Users/qiliu/Dropbox (VUMC)/DanBeachmap/geneinterest.txt"    #if "", no interesting gene list
#genelistfile<-"c:/Users/liuq6/Dropbox/single cell/genelist.txt" # the list of interesting genes 





```



```{r, echo=F}
cat("\n\n## cell type annotation parameters\n")

#cell type annotation

celltype_predictmethod="cta" # cta: cell activity; ora: over-represent

##specific marker database############

subtypefile<-"/Users/qiliu/Dropbox (VUMC)/DanBeachmap/singlecell_mouse_epithelium_signature.csv"
add_Subtype=FALSE ## add colorectal cancer subtype signature

markerfile<-"/Users/qiliu/Dropbox (Personal)/single cell/Rcode/Marker/PanglaoDB_markers_21_Oct_2019.tsv"

#these R will use markerfile, add_Subtype, species, so have to put at the end
source("/Users/qiliu/Dropbox (Personal)/single cell/Rcode/markerCode_filter.R")



knownmarkers<- c("COL1A1","COL1A2","DCN","TAGLN","MYH11","ATCA2","PECAM1","CDH5","CLDN5","CD68","C1QB","LYZ2","NKAIN4","LGFBP5","CD248","WIF1","CHAD","SULT1D1","CD3D","CD3G","NKG7","CD209A","H2-AB1","H2-EB1","KCNA1","PLP1","MYL1","CD79A")


if(species=="Mm") {
  

  knownmarkers<-paste0(substr(knownmarkers,1,1),substr(tolower(knownmarkers),2,nchar(knownmarkers)))
  
}

```




```{r,fig.width=15, fig.height=15,eval=analysis_individual,results="asis"}

cat("\n\n# QC and cell type identification in each sample\n")

object.list<-NULL

for (i in 1:nrow(SampleInfo)) {
  
 object<-preprocess(SampleInfo[i,],Cutoff[i,],Mtpattern,resolution, Remove_Mt_rRNA,celltype_predictmethod,transpose=transpose, analysis_individual=analysis_individual,Ensemblfile=Ensemblfile)
print(DoHeatmap(object,features=knownmarkers,slot="data"))
object.list<-c(object.list,object)
 
}


save(object.list,file=paste0(workdir,"/objectlist.Rdata"))


```



```{r,eval=analysis_integrated & batchcorrection}
#Use CCA to merge two conditions; depend on specific needs to decide to use it or not
 
  load(paste0(workdir,"objectlist.Rdata"))
  object.list<-object.list[c(2,3)]
  
   object.anchors <- FindIntegrationAnchors(object.list = object.list, dims = 1:30)
   object.integrated <- IntegrateData(anchorset = object.anchors, dims = 1:30)
  DefaultAssay(object.integrated) <- "integrated"
  object.integrated <- ScaleData(object.integrated, verbose = FALSE)
  object.integrated <- RunPCA(object.integrated, npcs = 30, verbose = FALSE)
  object.integrated <- FindNeighbors(object.integrated, dims = 1:30)
  object.integrated <-  FindClusters(object.integrated, resolution = 0.5)
  object.integrated <- RunUMAP(object.integrated, reduction = "pca", dims = 1:30)

DefaultAssay(object.integrated) <- "RNA"
medianexp<-getMedianExp(object.integrated)

predict_celltype<-ORA_celltype(medianexp,cellType,method=celltype_predictmethod)

#use the max_cta_score for each cluster


new.cluster.ids<-paste(levels(object.integrated@active.ident),rownames(predict_celltype$predict_result),sep="_")

object.integrated$cellcluster<-Idents(object.integrated)
names(new.cluster.ids) <- levels(object.integrated)
object.integrated <- RenameIdents(object.integrated, new.cluster.ids)
object.integrated$cellcluster_ann<-Idents(object.integrated)


integrated.markers <- FindAllMarkers(object.integrated, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)


object.integrated@misc$markers <- integrated.markers

save(object.integrated,file=paste0(workdir,"objectintegrated_batch.Rdata"))


```



```{r, eval=analysis_integrated & !batchcorrection,fig.width=12,fig.height=15}
cat("\n\n# cell clusters on integrated dataset")

 load(paste0(workdir,"objectlist.Rdata"))

 ## only work on two APC samples
 
  object.list<-object.list[c(2,3)]
  object.integrated<-merge(object.list[[1]], y = object.list[-1],  merge.data = TRUE)
  
  ##This is generate too many features### 
  #listvariables<-lapply(object.list,function(x){head(VariableFeatures(x),500)})
  
  #VariableFeatures(object.integrated)<-Reduce(union,listvariables)
  VariableFeatures(object.integrated)<-SelectIntegrationFeatures(object.list)
  rm(object.list)
 object.integrated <- ScaleData(object.integrated, verbose = FALSE)
 object.integrated <- RunPCA(object.integrated, npcs = 30, verbose = FALSE)
 object.integrated <- FindNeighbors(object.integrated, dims = 1:30)
object.integrated <-  FindClusters(object.integrated, resolution = 0.5)
object.integrated <- RunUMAP(object.integrated, reduction = "pca", dims = 1:30)


medianexp<-getMedianExp(object.integrated)

predict_celltype<-ORA_celltype(medianexp,cellType,method=celltype_predictmethod)

#use the max_cta_score for each cluster


new.cluster.ids<-paste(levels(object.integrated@active.ident),rownames(predict_celltype$predict_result),sep="_")

object.integrated$cellcluster<-Idents(object.integrated)
names(new.cluster.ids) <- levels(object.integrated)
object.integrated <- RenameIdents(object.integrated, new.cluster.ids)
object.integrated$cellcluster_ann<-Idents(object.integrated)


integrated.markers <- FindAllMarkers(object.integrated, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
object.integrated@misc$markers <- integrated.markers

save(object.integrated,file=paste0(workdir,"objectintegrated.Rdata"))


```




```{r,eval=Finalreport,results="asis",echo=F,fig.align="center",fig.cap=paste0("Fig.",1:ncol(clinicalInfo),"  QC by ",colnames(clinicalInfo))}

load(paste0(workdir,"objectintegrated_batch.Rdata"))
##QC across all features
  
cat("\n\n# Quality Check\n\n")
for (i in 1:ncol(clinicalInfo)) {
  var<-colnames(clinicalInfo)[i]
  cat("\n\n## QC across",var,"\n\n")
  plot1<- ggplot(data=data.frame(SampleId=unname(object.integrated[[var]]),nUMIs=object.integrated$nCount_RNA),aes(x=SampleId,y=nUMIs))+geom_boxplot()+scale_y_log10()+xlab(var)+theme(axis.text.x = element_text(angle = 90)) 

 plot2<- ggplot(data=data.frame(SampleId=unname(object.integrated[[var]]),nGenes=object.integrated$nFeature_RNA),aes(x=SampleId,y=nGenes))+geom_boxplot()+scale_y_log10()+xlab(var)+theme(axis.text.x = element_text(angle = 90)) 

 plot3<- ggplot(data=data.frame(SampleId=unname(object.integrated[[var]]),PercentMt=object.integrated$percent.mt),aes(x=SampleId,y=PercentMt))+geom_boxplot()+scale_y_log10()+xlab(var)+theme(axis.text.x = element_text(angle = 90)) 

 plot4<- ggplot(data=data.frame(SampleId=unname(object.integrated[[var]])),aes(x=SampleId))+geom_bar(stat="count")+xlab(var)+theme(axis.text.x = element_text(angle = 90))+ylab("nCells")
                
# plot4<-ggplot(data=data.frame(table(object.integrated[[var]],object.integrated$seurat_clusters)),aes(x=Var1,y=Freq, fill=Var2))+geom_bar(stat="identity")+ylab("nCells")+xlab(var)

print(CombinePlots(plots = list(plot1, plot2, plot3,plot4),rel_heights=c(1,1)) )

}


```





```{r,echo=F, eval=Finalreport,results="asis",fig.align="center",fig.cap=paste0("Fig.",2:5, c(" cell clusters"," cell clusters info"," cell clusters in UMAP"," cell clusters by sample")),fig.width=12}

cat("\n\n# Cell clusters on integrated dataset\n\n")

p1<-DimPlot(object.integrated, reduction = "pca",label=T,label.size=3)+labs(title = "PCA")+NoLegend()

p2<-DimPlot(object.integrated,reduction="umap",label=TRUE,label.size=3)+labs(title="UMAP")+NoLegend()
p3<- ggplot(data=data.frame(cluster=object.integrated@active.ident),aes(x=cluster))+geom_bar(stat="count")+theme(axis.text.x = element_text(angle = 90))+ylab("nCells")+ggtitle(paste0("total cells:",dim(object.integrated)[2]))
CombinePlots(plots = list(p1,p2,p3),ncol=3)

     plot1<- ggplot(data=data.frame(cluster=object.integrated$seurat_clusters,nUMI=object.integrated@meta.data$nCount_RNA),aes(x=cluster,y=nUMI))+geom_boxplot()+scale_y_log10()   
       plot2<- ggplot(data=data.frame(cluster=object.integrated$seurat_clusters,ngene=object.integrated@meta.data$nFeature_RNA),aes(x=cluster,y=ngene))+geom_boxplot()+scale_y_log10()
       plot3<-ggplot(data=data.frame(cluster=object.integrated$seurat_clusters,MtRNA=object.integrated@meta.data$percent.mt),aes(x=cluster,y=MtRNA))+geom_boxplot()  

print(CombinePlots(plots = list(plot1, plot2, plot3),ncol=3))

print(DimPlot(object.integrated,label=T))

tmp<-table(object.integrated@active.ident,object.integrated$SampleId)
barplot(t(tmp),beside=T,las=2,legend=colnames(tmp),cex.names=0.5)

```


```{r, eval=Finalreport,results="asis",echo=F,fig.width=15,fig.align="center",fig.cap=paste0("Fig.",6,"  Cell clusters group by ",colnames(clinicalInfo))}




cat("\n\n## Cell clusters group by Clinical Info\n\n")



for (i in 1:length(colnames(clinicalInfo))) {
print(DimPlot(object.integrated, reduction = "umap", group.by = colnames(clinicalInfo)[i], label = TRUE,label.size=3))
# print(DimPlot(object.integrated,reduction="umap",split=Condition[i],label=TRUE) ) 
}



```


```{r,eval=Finalreport,fig.width=15,fig.height=15,fig.align="center",fig.cap=paste0("Fig. ",7:11,c(" heatmap on scaled data"," heatmap on data"," heatmap for known markers", " heatmap of marke genes in DC", " heatmap of known markers  in DC ")),results="asis" }
cat("\n\n## Marker genes in each cluster\n\n")

integrated.markers <- object.integrated@misc$markers

topn <- integrated.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)

DoHeatmap(object.integrated, features = topn$gene,size=3,angle=90, assay="integrated")+theme(axis.text.y = element_text(size = 6))  
DoHeatmap(object.integrated, features = topn$gene,assay="RNA",slot="data",size=3,angle=90)+ theme(axis.text.y = element_text(size = 6))

DoHeatmap(object.integrated, features = knownmarkers, assay="RNA",slot="data",size=3,angle=90)


DoHeatmap(object.integrated[,object.integrated$seurat_clusters %in% c("3","5","6","10","13","14","16")], features = topn$gene,assay="RNA",slot="data",size=3,angle=90)+ theme(axis.text.y = element_text(size = 6))

DoHeatmap(object.integrated[,object.integrated$seurat_clusters %in% c("3","5","6","10","13","14","16")], features = knownmarkers, assay="RNA",slot="data",size=3,angle=90)


```

```{r,eval=Finalreport,fig.width=15,fig.height=7,fig.align="center",results="asis" }
cat("\n\n## UMAP and VlnPlot for marker genes\n\n")


ngene<-length(topn$gene)
for (i in 1:(ngene/4)) {
print(FeaturePlot(object.integrated,features=topn$gene[(4*(i-1)+1):(4*i)],ncol=4))

}

print(FeaturePlot(object.integrated,features=topn$gene[(4*i+1):ngene]))

for (i in 1:(ngene/2)) {
print(VlnPlot(object.integrated,features=topn$gene[(2*(i-1)+1):(2*i)]))

}

print(VlnPlot(object.integrated,features=topn$gene[(2*i+1):ngene]))



```

```{r,eval=Finalreport,fig.width=15,fig.height=7,fig.align="center",results="asis" }
cat("\n\n## UMAP and VlnPlot for known markers\n\n")

knownmarkers<-intersect(knownmarkers,rownames(object.integrated))

ngene<-length(knownmarkers)
for (i in 1:(ngene/4)) {
print(FeaturePlot(object.integrated,features=knownmarkers[(4*(i-1)+1):(4*i)],ncol=4))

}

print(FeaturePlot(object.integrated,features=knownmarkers[(4*i+1):ngene]))

for (i in 1:(ngene/2)) {
print(VlnPlot(object.integrated,features=knownmarkers[(2*(i-1)+1):(2*i)]))

}

print(VlnPlot(object.integrated,features=knownmarkers[(2*i+1):ngene]))



```


```{r,eval=F,echo=FALSE,results="asis"}
cat("\n\n# The cell type composition across samples\n\n")


library(permute)
library(GUniFrac)

hvgdata_pca<-Embeddings(object.integrated, reduction = "pca")
memb<-object.integrated@active.ident



cluster<-unique(memb)
ncluster<-length(cluster)
              
        
       
    ##generate the table count for each cell type
    count.table<-table(object.integrated@meta.data$SampleId,object.integrated@active.ident)
    prop_celltype<-count.table/apply(count.table,1,sum)
    prop_celltype<-as.matrix(unclass(prop_celltype))
    
    
    cent <- NULL
     
    for(k in 1:ncluster){
        cent <- rbind(cent, colMeans(hvgdata_pca[memb == cluster[k], , drop = FALSE]))
        
    }
    rownames(cent)<-cluster
    hc1 <- hclust(dist(cent), method = "ave", members = table(memb))
    tree1<-as.phylo(hc1)
    ##calculate the distance
    unifracs <- GUniFrac(count.table, tree1, alpha=c(0, 0.5, 1))$unifracs
    dist.obs<-unifracs[, , "d_1"]
    rownames(dist.obs)<-colnames(dist.obs)<-rownames(count.table)
    
 
if (nrow(SampleInfo)>2) {
pcoa_result<-pcoa(dist.obs)

}
    
    
```


```{r,eval=F,echo=FALSE}
   ##permuate the samples, calculate the pvalue
    nperm=1000
    nsample<-length(unique(group))
    ncell<- length(group)
    control <- how(nperm = nperm, within = Within(type = "free"))
    t.permu <- numeric(length = control$nperm) +1
    dis.permu.array<-array(0,dim=c(nsample,nsample,nperm))
    for(numPer in seq_along(t.permu)) {
        
        want <- permute(numPer, ncell, control)
        group.permu<-group[want]
        
        ## calculate distance
        count.table.perm<-table(group.permu,memb)
        colnames(count.table.perm)<-1:ncluster
        unifracs.perm <- GUniFrac(count.table.perm, tree1, alpha=c(0, 0.5, 1))$unifracs
        dis.permu.array[,,numPer]<- unifracs.perm[, , "d_1"]
    }
    
    pvalue<-matrix(apply(apply(dis.permu.array,3,">=",dist.obs),1,sum),nrow=nsample,byrow=F)/nperm
    rownames(pvalue)<-colnames(pvalue)<-rownames(count.table)


```

```{r,eval=F,echo=FALSE,results="asis",fig.align="center",fig.cap=paste0("Fig.", (2*ncol(clinicalInfo)+5):((2*ncol(clinicalInfo)+10)),c(" barplot of cell type composition"," cell type distribution in Histo"," cell type distribution in Location", " cell type distribution"," heatmap of cell type composition"," PCoA results")),fig.width=12,fig.height=10}    

#match the orignial order
#prop_celltype<-prop_celltype[match(SampleInfo$SampleId),rownames(prop_celltype)),]
SampleInfo<-subset(SampleInfo,select = -c(countfile))
tmp_short<-merge(SampleInfo,prop_celltype,by.x=1,by.y="row.names")

tmp<-melt(tmp_short,id.vars=colnames(SampleInfo),value.name="prop",variable.name="CellCluster")

cols<-sample(grDevices::colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)],dim(count.table)[2])

ggplot(tmp,aes(x=SampleId,y=prop,fill=as.factor(CellCluster)))+geom_bar(stat="identity")+ scale_fill_manual(values = cols)+theme(axis.text.x = element_text(angle = 90))

##
tmp$Histo <- factor(tmp$Histo , levels=c("Moderate", "Mild", "Quiescent", "Normal"))
###
ggplot(tmp, aes(x=tmp[,clinicalfeature[1]], y=prop,col=tmp[,clinicalfeature[1]])) +   geom_jitter(width=0.25) +facet_wrap(~CellCluster,scales="free")+geom_violin()+theme(axis.text.x = element_text(angle = 90)) 
ggplot(tmp, aes(x=tmp[,clinicalfeature[2]], y=prop,col=tmp[,clinicalfeature[2]])) +   geom_jitter(width=0.25) +facet_wrap(~CellCluster,scales="free")+geom_violin()+theme(axis.text.x = element_text(angle = 90)) 

ggplot(tmp, aes(x=tmp[,clinicalfeature[1]], y=prop, color=Location,shape=tmp[,clinicalfeature[2]])) +  geom_jitter(width=0.25) +facet_wrap(~CellCluster,scales="free")+geom_violin()+theme(axis.text.x = element_text(angle = 90))  

####
rownames(SampleInfo)<-SampleInfo$SampleId


x<-t(prop_celltype)
annotation_col= data.frame(SampleInfo[,clinicalfeature])
names(annotation_col)<-colnames(SampleInfo)[clinicalfeature]
pheatmap(x,annotation_col = annotation_col,border_color="white",cellwidth=10,cellheight=10,angle_col= 45,fontsize=8,fontsize_row=10,fontsize_col=10)
  
  


if (nrow(SampleInfo)>2) {

SampleInfo<-merge(SampleInfo,prop_celltype,by.x=1,by.y="row.names")

SampleInfo<- merge(SampleInfo,pcoa_result$vectors[,1:2],by.x=1,by.y="row.names")

if (length(clinicalfeature)>=2) {
ggplot(SampleInfo,aes(x=Axis.1,y=Axis.2,col=SampleInfo[,clinicalfeature[1]],shape=SampleInfo[,clinicalfeature[2]]))+geom_point()+theme_classic()+labs(title="PCoA",x="PCoa1",y="PCoa2")+geom_text(label=SampleInfo$SampleId,nudge_x = 0.01, nudge_y = 0.005) } else {ggplot(SampleInfo,aes(x=Axis.1,y=Axis.2,col=SampleInfo[,clinicalfeature[1]]))+geom_point()+theme_classic()+labs(title="PCoA",x="PCoa1",y="PCoa2")+geom_text(label=SampleInfo$SampleId,nudge_x = 0.01, nudge_y = 0.005)}
#ggplot(SampleInfo,aes(x=Axis.1,y=Axis.2,shape=colnames(SampleInfo)[3],col=colnames(SampleInfo)[7]))+geom_point()+geom_text(label=SampleInfo$SampleId)+theme_classic()+scale_shape_manual(values=c(17,16))+scale_size_manual(values=(c(6,4)))+labs(title="PCoA",x="PCoa1",y="PCoa2")+facet_wrap(~PatientID)

} 



```



```{r,eval=F,echo=F,results="asis",fig.align="center",fig.cap=paste0("Fig.", (2*ncol(clinicalInfo)+11):((2*ncol(clinicalInfo)+13)),c(" barplot of cell type composition in TI samples"," heatmap of cell type composition in TI samples"," PCoA results of TI samples"))}
##this is specific analysis for certain projects
cat("\n\n# Focus on TI\n\n")
SampleInfo_TI<-SampleInfo[SampleInfo$Sampletype=="Terminal Ileum",]
rownames(SampleInfo_TI)<-SampleInfo_TI$SampleId

prop_celltype_TI<-prop_celltype[rownames(prop_celltype)%in%SampleInfo$SampleId[SampleInfo$Sampletype=="Terminal Ileum"],]



prop_celltype_TI<-SampleInfo_TI[,(dim(clinicalInfo)[2]+2):(dim(SampleInfo_TI)[2]-2)]
rownames(prop_celltype_TI)<-SampleInfo_TI$SampleId

tmp<-melt(as.matrix(prop_celltype_TI),id=1,varnames=c("id","group"),value.name="prop")
tmp$id<-paste0("S",tmp$id)

ggplot(tmp,aes(x=id,y=prop,fill=as.factor(group)))+geom_bar(stat="identity")+ scale_fill_manual(values = cols)+ theme(axis.text.x = element_text(angle = 45))

annotation_col= data.frame(SampleInfo_TI[,clinicalfeature])
names(annotation_col)<-colnames(SampleInfo_TI)[clinicalfeature]
pheatmap(t(prop_celltype_TI),annotation_col = SampleInfo_TI[,clinicalfeature],border_color="white",cellwidth=10,cellheight=10,angle_col= 45,fontsize=8,fontsize_row=10,fontsize_col=10)


dist.obs.TI<-dist.obs[rownames(dist.obs)%in%rownames(prop_celltype_TI),rownames(dist.obs)%in%rownames(prop_celltype_TI)]

pcoa_result_TI<-pcoa(dist.obs.TI)
ggplot(as.data.frame(pcoa_result_TI$vectors),aes(x=Axis.1,y=Axis.2,col=SampleInfo_TI[,clinicalfeature[1]]))+geom_point()+theme_classic()+labs(title="PCoA",x="PCoa1",y="PCoa2")+geom_text(label=rownames(pcoa_result_TI$vectors),nudge_x = 0.01, nudge_y = 0.005)

```























```{r, eval=F,fig.width=15,fig.height=15,results="asis"}
cat("#\n\n differential between conditions\n\n")

DefaultAssay(object.integrated) <- "RNA"
object.integrated <- ScaleData(object.integrated, verbose = FALSE)

## differential expression between KO and WT on the dataset before integration


 object.integrated$celltype<-Idents(object.integrated)
 Idents(object.integrated)<-"Condition"
 
 n_diff<-FindMarkers(object.integrated,ident.1=levels(object.integrated@active.ident)[1],ident.2=levels(object.integrated@active.ident)[2],verbose=FALSE)
 head(n_diff,15)
 DoHeatmap(object.integrated, features = c("Smad4",rownames(n_diff)[1:50])) + NoLegend()




```




```{r, eval=F, fig.width=15,fig.height=15}
#use the max_cta_score for each cluster


new.cluster.ids<-paste(levels(object.integrated@active.ident),rownames(predict_celltype$predict_result),sep="_")

names(new.cluster.ids) <- levels(object.integrated)
object.integrated <- RenameIdents(object.integrated, new.cluster.ids)
object.integrated$celltype_ann<-Idents(object.integrated)
DimPlot(object.integrated, reduction = "umap",label=T,label.size=4)
DoHeatmap(object.integrated, features = top10$gene) + NoLegend()

```

```{r,eval=F, fig.width=15, fig.height=15}

Doheatmap_cellmarker_cellType(object.integrated,top10,cellType,predict_celltype)
```









```{r,eval=F,fig.width=15, fig.height=15,results="asis"}
cat("#\n\n Core response genes\n\n")

Idents(object.integrated)<-"Condition"
core.response<-FindConservedMarkers(object.integrated,ident.1=levels(object.integrated@active.ident)[1],grouping.var="celltype",verbose=FALSE)
head(core.response)
write.csv(core.response,file="/Users/qiliu/Dropbox (VUMC)/DanBeachmap/coreresponse.csv")
Idents(object.integrated)<-"celltype"
DotPlot(object.integrated,features=rownames(head(core.response,15)),cols=c("blue","red"),dot.scale=8,split.by="Condition")+RotatedAxis()


```



```{r,eval=F,fig.width=15, fig.height=15,results="asis"}
cat("\n\n# cell type-specific response genes\n\n")

Idents(object.integrated)<-paste(object.integrated$celltype,object.integrated$Condition,sep="_")
clusternames<-levels(object.integrated$celltype)
groupnames<-unique(object.integrated$Condition)
specific.response<-NULL
for (i in 1:length(clusternames)){
  each.specific.response<-FindMarkers(object.integrated,ident.1=paste(clusternames[i],groupnames[1],sep="_"),ident.2=paste(clusternames[i],groupnames[2],sep="_"),verbose=FALSE)
   head(each.specific.response,20)
   tmp<-each.specific.response
   rownames(tmp)<-c()
   specific.response<-rbind(specific.response,cbind(gene=rownames(each.specific.response),tmp,cluster=clusternames[i]))
   genenames<-rownames(each.specific.response)[!rownames(each.specific.response)%in%rownames(core.response)]
   
  print(FeaturePlot(object.integrated, features = genenames[1:4], split.by = "Condition", max.cutoff = 3, 
    cols = c("grey", "red")))
  
}

  write.csv(specific.response,file="/Users/qiliu/Dropbox (VUMC)/DanBeachmap/specificresponse.csv",row.names = F)
```






```{r,eval=F,fig.width=15, fig.height=15,results="asis"}
cat("\n\n# Plot for interesting genes\n\n")
if (file.exists(genelistfile)) {
genelist<-data.frame(fread(genelistfile,header=F))[,1]

genelist<-genelist[genelist%in%rownames(GetAssayData(object.integrated,slot="data"))]
for (i in seq(1,length(genelist),4)){
 print(FeaturePlot(object.integrated, features = genelist[i:min(length(genelist),(i+3))], split.by = "Condition", max.cutoff = 3, 
    cols = c("grey", "red")))
 
}

integrated.markers[rownames(integrated.markers)%in%genelist,]
}


Idents(object.integrated)<-"celltype"
VlnPlot(object.integrated, features = genelist, split.by = "Condition", group.by = "celltype", 
    pt.size = 0, combine = FALSE)

```





```{r, eval=F,results="asis"}
library(monocle3)
cat("\n\n# Trajectory path\n\n")
for (i in 1:length(object.list)) {
object.path<-object.list[[i]]
cds<-new_cell_data_set(GetAssayData(object.path,slot="data"),cell_metadata = object.path@meta.data)

cds<-preprocess_cds(cds,norm_method = "none")
cds<-reduce_dimension(cds)
plot_cells(cds,label_groups_by_cluster = FALSE,color_cells_by = "celltype_each")
cds <- cluster_cells(cds)
cds <- learn_graph(cds)
print(plot_cells(cds,
           color_cells_by = "celltype_each",
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE))

#cds_gene<-graph_test(cds, neighbor_graph="principal_graph", cores=4)
#cds_deg_ids <- row.names(subset(cds_gene, q_value < 0.05))
#plot_cells(cds, genes=c("CFTR","ZG16","REG4","TCF7L2"),
#           show_trajectory_graph=FALSE,
#           label_cell_groups=FALSE,
#           label_leaves=FALSE)

}

object.path<-object.integrated
cds<-new_cell_data_set(GetAssayData(object.path,slot="data"),cell_metadata = object.path@meta.data)

cds<-preprocess_cds(cds,norm_method = "none")
cds<-reduce_dimension(cds)
plot_cells(cds,label_groups_by_cluster = FALSE,color_cells_by = "celltype_each")
cds <- cluster_cells(cds)
cds <- learn_graph(cds)
plot_cells(cds,
           color_cells_by = "celltype_each",
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE)

```