# Differential Gene Report

<br>

```{r,echo=FALSE}
options(width = 3000,warning=FALSE)

interestedModuleInd=find_module_folder(files,pattern="limma")
diffGeneFiles=files[interestedModuleInd,]

```

## Overall Gene Expression

```{r diffGeneBoxplotRaw, results="asis", echo=FALSE, fig.align="center",fig.cap=figRef("gene_exp_boxplot_raw", "Gene Expression distribution before normlization")}
boxplotFigInd=grep(".RawBoxplot.png$",diffGeneFiles[,1])
check_and_include_graphics(diffGeneFiles[boxplotFigInd,1])
```

```{r diffGeneBoxplotNormlized, results="asis", echo=FALSE, fig.align="center",fig.cap=figRef("gene_exp_norm_boxplot", "Gene Expression distribution after normlization")}
boxplotFigInd=grep(".NormlizationBoxplot.png$",diffGeneFiles[,1])
check_and_include_graphics(diffGeneFiles[boxplotFigInd,1])
```

```{r diffGeneHeatmap, results="asis", echo=FALSE, fig.align="center",fig.cap=figRef("gene_exp_heatmnap", "Gene Expression Heatmap")}
boxplotFigInd=grep(".Heatmap.png$",diffGeneFiles[,1])
check_and_include_graphics(diffGeneFiles[boxplotFigInd,1])
```

## Differential Gene Detection

### Differential Result Summary

```{r, results="asis", echo=FALSE}
diffDesignTableFileInd=grep(".OverallDesignTable.csv$",diffGeneFiles[,1])
diffDesignTable=read.csv(diffGeneFiles[diffDesignTableFileInd,1],row.names=1, check.names=F)
print(kable(diffDesignTable,caption = tabRef("design_table","Differential Detection Design Table")) %>%	
  kable_styling() %>%
  htmltools::HTML())
```

```{r, results="asis", echo=FALSE}
ComparisonsTableFileInd=grep(".ComparisonsTable.csv$",diffGeneFiles[,1])
ComparisonsTableFile=read.csv(diffGeneFiles[ComparisonsTableFileInd,1],row.names=1)
print(kable(ComparisonsTableFile,caption = tabRef("comparisons_table","Differential Detection Comparisons"))%>%	
  kable_styling() %>%
  htmltools::HTML())
```


```{r, results="asis", echo=FALSE}
diffCriteriaFileInd=grep(".DiffCriteria.csv$",diffGeneFiles[,1])
diffCriteriaTable=read.csv(diffGeneFiles[diffCriteriaFileInd,1])
#cat("Differential expression analysis criteria:  fold change >= ", rnaseqOptions["DE_fold_change", 1], " and FDR <= ", rnaseqOptions["DE_pvalue", 1], "\n\n")
print(kable(diffCriteriaTable,caption = tabRef("DEG_Criterias","DEG Criterias"))%>%	
  kable_styling() %>%
  htmltools::HTML())
```


```{r diffGeneTableNumber, results="asis", echo=FALSE, fig.align="center"}

diffGeneTableInd=grep("_Diff_sig.csv$",files[,1])


if (length(diffGeneTableInd)>0) {
  DESeq2_files=files[diffGeneTableInd,,drop=FALSE]
  
  #summary Table
  DESeq2_list<-list()
  for (i in 1:nrow(DESeq2_files)){
    DESeq2_list[[i]]<-read.csv(DESeq2_files[i,1], header=TRUE, check.names=F)
  }
  output_table3<-data.frame(c(gsub("_DESeq2_sig.csv","",basename(DESeq2_files[,1]))), c(unlist(lapply(DESeq2_list, nrow))))
  colnames(output_table3)<-c("Comparison", "Number of DEGs")
  print(kable(output_table3,caption = tabRef("DEG_number","DEG numbers"))%>%	
  kable_styling() %>%
  htmltools::HTML())
}


```

### Volcano plot
```{r diffGeneVolcanoPlot, results="asis", echo=FALSE, fig.align="center",fig.cap=figRef("diff_gene_VolcanoPlot", "Volcano Plot of Differential Gene Detection")}
boxplotFigInd=grep(".VolcanoPlot.png$",diffGeneFiles[,1])
check_and_include_graphics(diffGeneFiles[boxplotFigInd,1])
```

### Venn plot

```{r, results="asis", echo=FALSE, fig.align="center",fig.cap=figRef("diff_gene_Venn", "Venn Plot of Differential Gene Detection")}
vennFigInd=grep(".Venn.png$",diffGeneFiles[,1])
check_and_include_graphics(diffGeneFiles[vennFigInd,1])
```


### Top Differential Genes

```{r diffGeneTable, results="asis", echo=FALSE, fig.align="center"}

if (length(diffGeneTableInd)>0) {
  DESeq2_files=files[diffGeneTableInd,,drop=FALSE]

  for (i in 1:nrow(DESeq2_files)){
    comparisonName<-DESeq2_files[i,2]
    deseq2<-read.csv(DESeq2_files[i,1], header=TRUE, row.names=1, check.names=F)
    if(nrow(deseq2) > 0){
      maxrow<-min(nrow(deseq2), 10)
#      cnames<-c("baseMean", "log2FoldChange","lfcSE", "stat", "pvalue", "padj", "FoldChange")
      cnames<-c("AveExpr", "logFC", "t", "P.Value", "adj.P.Val")
      cnames=c(intersect(c("GeneSymbol","entry_name"),colnames(deseq2)),cnames)
      deseq2table<-deseq2[c(1:maxrow), cnames]
      if (any(grepl("[A-Za-z]",row.names(deseq2table)))) {
        keepRowNames=1
      } else {
        keepRowNames=0
      }
      print(kable(deseq2table, row.names=keepRowNames, caption= tabRef(comparisonName, paste0(comparisonName, " top ", maxrow, " differential expressed genes")))%>% 
          kable_styling() %>%
          row_spec(which(deseq2table$logFC > 0), color = "black", background = "bisque") %>% 
          row_spec(which(deseq2table$logFC < 0), color = "black", background = "honeydew") %>%
          htmltools::HTML())
    }
  }
}


```

