---
title: "ChIP-Seq Analysis Report"
date: "`r format(Sys.time())`"
output: 
  rmdformats::readthedown:
    toc_depth: 3
    number_sections: true
---

```{css, echo=FALSE}
#content{
    max-width:1920px;
}
```
  
```{r setup, include=FALSE}
library(XML)
library(rlist)
library(stringr)
library(kableExtra)
options(bitmapType='cairo')

source('reportFunctions.R')
source('Pipeline.R')

files<-read.table("fileList1.txt", header=FALSE, as.is=TRUE)
files<-split(files$V1, files$V2)

chipseqOptions_df<-read.table("fileList2.txt", sep="\t", header=FALSE, as.is=TRUE)
chipseqOptions<-split(chipseqOptions_df$V1, chipseqOptions_df$V2)

has_fastqc_raw<-any(grepl("fastqc_raw_", names(files)))
has_fastqc_post<-any(grepl("fastqc_post_trim_", names(files)))
has_bowtie2 <- any(grepl("bowtie2_", names(files)))
has_bowtie2_cleanbam <- any(grepl("bowtie2_cleanbam_", names(files)))
has_chipqc <- "chipqc_html" %in% names(files)
has_homer<-any(grepl("homer_", names(files)))

has_peak_count<-"peak_count" %in% names(files)

homer<-chipseqOptions$homer_result
macs2<-chipseqOptions$macs2_result

if(has_chipqc) {
  chipqc_html = files$chipqc_html
  has_chipqc=file.exists(chipqc_html)
  chipqc_dir=dirname(chipqc_html)
  chp_file=file.path(chipqc_dir, "CoverageHistogramPlot.png")
  gfe_file=file.path(chipqc_dir, "GenomicFeatureEnrichment.png")
  pch_file=file.path(chipqc_dir, "PeakCorHeatmap.png")
  pca_file=file.path(chipqc_dir, "PeakPCA.png")
  pp_file=file.path(chipqc_dir, "PeakProfile.png")
  rap_file=file.path(chipqc_dir, "Rap.png")
  rip_file=file.path(chipqc_dir, "Rip.png")
}else{
  chp_file=""
  gfe_file=""
  pch_file=""
  pca_file=""
  pp_file=""
  rap_file=""
  rip_file=""
}

project_dir <- dir("../..", full.names=T)

report <- project_dir[grep("report", project_dir)]

has_macs2 <- ifelse(is.null(macs2), FALSE, file.exists(macs2))
has_report <- FALSE

has_diffbind<-"diffbind_result" %in% names(chipseqOptions)
if(has_diffbind) {
  diffbind <- chipseqOptions$diffbind_result
}

getFigure <- function(category, description){
  return(paste0("```{r,echo=FALSE,results='asis', fig.align='center', fig.cap=figRef('", category, "', '",description, "')}\n",
"  check_and_include_graphics(files['", category, "',1])\n```\n"))
}

```
 
```{r copyfiles, include=FALSE, eval=has_macs2, results='asis', message=FALSE, warning=FALSE}
#system("mkdir ChIP_seq_results")
#system("mkdir ChIP_seq_results/diffbind")
#files_diff <- c(list.files(paste0(diffbind, "/result"), pattern=".png", full.names=T, recursive=T),
#list.files(paste0(diffbind, "/result"), pattern=".sig\\.", full.names=T, recursive=T))
#file.copy(from = files_diff, to = "ChIP_seq_results/diffbind/")
```


<br>
  
# Summary

```{r summary, results="asis", echo=FALSE} 
cat("This report includes quality check for raw sequencing data (section 2), reads mapping, duplication level, reads enrichment in genomic features, ChIP signal distribution and structure (section 3), and the correlation between samples (section 4).")
cat(" The results of differential analysis of peak intensity are summarized in section 5, ")
cat("and the results folder structure are described in section 6.")
cat("\n")
```

<br>

# Sequencing Quality
  
```{r fastqc_raw_per_base_sequence_quality,echo=FALSE,results='asis', include = has_fastqc_raw, eval = has_fastqc_raw, fig.align="center", fig.cap=figRef("fastqc_raw_per_base_sequence_quality", "The mean quality scores of sequencing reads in each position", trunk.eval=file.exists(files$fastqc_raw_per_base_sequence_quality))}
check_and_include_graphics(files$fastqc_raw_per_base_sequence_quality)
```

<br>

```{r fastqc_raw_per_sequence_gc_content,echo=FALSE,results='asis', include = has_fastqc_raw, eval = has_fastqc_raw, fig.align="center", fig.cap=figRef("fastqc_raw_per_sequence_gc_content", "The average GC content of sequencing reads", trunk.eval=file.exists(files$fastqc_raw_per_sequence_gc_content))}
check_and_include_graphics(files$fastqc_raw_per_sequence_gc_content)
```

<br>
  
```{r fastqc_raw_adapter_content,echo=FALSE,results='asis', include = has_fastqc_raw, eval = has_fastqc_raw, fig.align="center", fig.cap=figRef("fastqc_raw_adapter_content", "Adapter content of sequencing reads", trunk.eval=file.exists(files$fastqc_raw_adapter_content))}
check_and_include_graphics(files$fastqc_raw_adapter_content)
```

<br>
  
```{r fastqc_post_trim_adapter_content,echo=FALSE,results='asis', include = has_fastqc_post, eval = has_fastqc_post, fig.align="center", fig.cap=figRef("fastqc_post_trim_adapter_content", "Adapter content of sequencing reads after cutadapt", trunk.eval=file.exists(files$fastqc_post_trim_adapter_content))}
check_and_include_graphics(files$fastqc_post_trim_adapter_content)
```

<br>
  
# Mapping quality

## Mapping, filtering and duplication rate
```{r bowtie2_reads,echo=FALSE,results='asis', eval = has_bowtie2, warning=FALSE, fig.align="center", fig.cap=figRef("bowtie2_summary_reads", "The statistics of ChIP-seq mapping results by reads")}
check_and_include_graphics(files$bowtie2_summary_reads)
```

```{r bowtie2_chromosome,echo=FALSE,results='asis', eval = has_bowtie2, warning=FALSE, fig.align="center", fig.cap=figRef("bowtie2_summary_chromosome", "The statistics of ChIP-seq mapping results by chromsome")}
check_and_include_graphics(files$bowtie2_summary_chromosome)
```

```{r bowtie2_cleanbam_chromosome,echo=FALSE,results='asis', eval = has_bowtie2_cleanbam, warning=FALSE, fig.align="center", fig.cap=figRef("bowtie2_cleanbam_chromosome", "The statistics of ChIP-seq mapping results by chromsome after cleaning bam files")}
check_and_include_graphics(files$bowtie2_cleanbam_summary_chromosome)
```


<br>

```{r ChipQC_summary_table, echo=FALSE, results='asis', eval=has_chipqc}
QCtables <- readHTMLTable(chipqc_html)
QCtables <- list.clean(QCtables, fun = is.null, recursive = FALSE)
QCtable <- QCtables[[2]]
colnames(QCtable) <- gsub('\n','',colnames(QCtable))
print(kable(QCtable, caption=tabRef("mapSummary", "Summary of ChIP-seq filtering and quality metrics."))%>%
    kable_styling() %>%
    htmltools::HTML())
```

<br>

* **ID** - Unique sample ID.  
* **Tissue/Factor/Condition** - Metadata associated to sample.  
* **Replicate** - Number of replicate within sample group  
* **Reads** - Number of sample reads within analysed chromosomes.  
* **Dup%** - Percentage of MapQ filter passing reads marked as duplicates  
* **FragLen** - Estimated fragment length by cross-coverage method  
* **SSD** - SSD score (htSeqTools)  
* **FragLenCC** - Cross-Coverage score at the fragment length  
* **RelativeCC** - Cross-coverage score at the fragment length over Cross-coverage at the read length  
* **RIP%** - Percentage of reads wthin peaks  
* **RIBL%** - Percentage of reads wthin Blacklist regions  

<br>

```{r genomic_features,echo=FALSE,warning=FALSE, results='asis', eval=file.exists(gfe_file), fig.align="center", fig.cap=figRef("feature_enrichment", "Heatmap of log2 enrichment of reads in genomic features")}
cat("## Enrichment of reads in genomic features\n")
include_graphics(gfe_file)
```

<br>

```{r CrossCoverage,echo=FALSE,warning=FALSE, results='asis', eval=file.exists(chp_file), fig.align="center", fig.cap=figRef("cross_coverage", "Plot of CrossCoverage score after successive strand shifts")}
cat("## ChIP signal distribution and structure\n")
include_graphics(chp_file)
```

<br>

# Peak calling and correlation between samples

## Summary of peak calling

```{r peak_calling_summary, echo=FALSE, warning=FALSE, results='asis', eval=has_macs2}

peaks_xls <- list.files(paste0(macs2, "/result"), recursive=T, pattern="peaks.xls", full.names=T)
if(has_homer){
  peaks_annot <- list.files(paste0(homer, "/result"), recursive=T, pattern=".bed.annotation.stats$", full.names=T)
}
sample_names <- gsub("_peaks.xls", "", basename(peaks_xls))

df <- data.frame()
for(i in 1:length(peaks_xls)){
  lines <- readLines(peaks_xls[i], n=29)
  df[i,1] <- sample_names[i]
  df[i,2] <- word(lines[grep("# fragment size is determined as", lines)], 7)
  df[i,3] <- word(lines[grep("# d = ", lines)], 4)
  qvalue_index = grep("# qvalue cutoff for broad/weak regions", lines)
  if(length(qvalue_index) == 0){
    qvalue_index = grep("# qvalue cutoff for narrow/strong regions", lines)
  }
  if(length(qvalue_index) == 0){#pvalue
    df[i,4] <- paste0("<= ", word(lines[11], 5))
  }else{
    df[i,4] <- paste0("<= ", word(lines[qvalue_index], 8))
  }
  df[i,5] <- dim(fread(peaks_xls[i], data.table = F))[1]
  if(has_homer){
    stats <- fread(peaks_annot[i], data.table=F)
    df[i,6] <- stats[9,4]
    df[i,7] <- stats[10,4]
  }
}
if(has_homer){
  colnames(df) <- c("Sample", "Read length", "Fragment length", "qvalue", "Peak number", 
				"log2 enrichment promoter", "log2 enrichment 5UTR")
}else{
  colnames(df) <- c("Sample", "Read length", "Fragment length", "qvalue", "Peak number")
}

print(kable(df, caption=tabRef("peakSummary", "Summary of peak calling and enrichment")) %>%
    kable_styling() %>%
    htmltools::HTML())
```

```{r peak_count, echo=FALSE, results='asis', warning=F, eval=has_peak_count}
df<-read.table(files$peak_count, sep="\t", header=T)
print(kable(df, caption=tabRef("peakSummary", "Summary of peak calling")) %>%
    kable_styling() %>%
    htmltools::HTML())
```
<br>

```{r homer_motif, echo=FALSE, results='asis', warning=F, eval=has_homer}
cat("## Motif enrichment\n")
dirs <- paste0(list.dirs(paste0(homer, "/result"), full.names=T, recursive=F), "/knownResults")

res_table <- data.frame()
i=1
for (i in 1:length(dirs)){
  curdir = dirs[i]
  knownRes_file <- paste0(curdir, ".txt")
  res_txt <- fread(knownRes_file, header=T, nrows=5, data.table=F)
  sample=basename(dirname(curdir))
  res_txt$Sample=sample
  res_txt$Motif=""

  for (j in 1:nrow(res_txt)) {
    svgfile = paste0(curdir, "/known", j, ".logo.svg")
    pngfile = paste0(curdir, "/known", j, ".logo.png")
    if (file.exists(svgfile)){
      if(!file.exists(pngfile)){
        x <- paste0("convert -density 1200 -resize 200x200 ", svgfile, " ", pngfile) 
      #cat(x, "\n")
        system(x)
      }
      res_txt$Motif[j]=paste0("<img src= ", pngfile, " >")
    }
  }

  res_table <- rbind(res_table, res_txt[,c("Motif Name", "P-value", "q-value (Benjamini)", "Sample", "Motif")])
}

tab <- tibble::tibble(Sample = res_table$Sample,
                      Name = gsub("/Homer","",res_table[,1]),
                      Motif = noquote(res_table$Motif),
                      Pvalue= res_table[,2], FDR = res_table[,3])
print(kable(tab, caption= tabRef("homer_table", "Homer annotation for peaks"), escape = F) %>%
    kable_styling() %>%
    htmltools::HTML())
```
<br>

```{r,echo=FALSE,results='asis', eval=file.exists(pp_file), fig.align="center", fig.cap=figRef("peak_profile", "Plot of the average signal profile across peaks")}
cat("## Peak profile and ChIP enrichment\n")
include_graphics(pp_file)
```

```{r,echo=FALSE,results='asis', eval=file.exists(rap_file), fig.align="center", fig.cap=figRef("rap", "Barplot of the percentage number of reads in peaks")}
include_graphics(rap_file)
```

```{r,echo=FALSE,results='asis', eval=file.exists(rip_file), fig.align="center", fig.cap=figRef("rip", "Density plot of the number of reads in peaks")}
include_graphics(rip_file)
```

```{r Correlation,echo=FALSE,results='asis', eval=file.exists(pch_file), fig.align="center", fig.cap=figRef("peak_correlation", "Plot of correlation between peaksets")}
cat("## Correlation between samples\n")
include_graphics(pch_file)
cat("\n* Replicate samples of high quality can be expected to cluster together in the heatmap and be spatially grouped within the PCA plot.\n\n")
```

```{r PCA,echo=FALSE,results='asis', eval=file.exists(pca_file), fig.align="center", fig.cap=figRef("pca", "PCA of peaksets")}
include_graphics(pca_file)
```

```{r diffbind_1, results="asis", echo=FALSE, eval=has_diffbind} 
#if(has_diffbind){
  cat("## Differential binding affinity analysis\n<br>\n\n### Binding sites overlaps\n<br>\n")
#}
```

```{r bindiff_Venn,echo=FALSE,results='asis', eval=has_diffbind, fig.align="center", fig.cap=figRef("diffbindVenn", "Binding sites overlaps across replicates"), out.width="30%", out.height="15%",fig.show='hold'}
diff_files <- list.files(paste0(diffbind, "/result"), pattern=".pdf", full.names=T, recursive=T)

for (i in 1:length(diff_files)){
    x <- paste0("convert ", diff_files[i], " ", gsub("pdf", "png", diff_files[i]))
    system(x)
}

venn <- list.files(paste0(diffbind, "/result"), pattern=".png", full.names=T, recursive=T)
venn <- venn[-grep("condition.png", venn)]

include_graphics(venn)
```

<br>

```{r bindiff_Venn2,echo=FALSE,results='asis', eval=has_diffbind, fig.align="center", fig.cap=figRef("diffbindVennOverlap", "Binding sites overlaps across conditions"), out.width="30%", out.height="15%",fig.show='hold'}
condition_file = list.files(paste0(diffbind, "/result"), pattern="Overlap-condition.png", full.names=T, recursive=T)
include_graphics(condition_file)
```

<br>

```{r bindiff_2, eval=has_diffbind, results="asis", echo=FALSE} 
cat("### Dysregulation between conditions\n<br>\n")
sig_table <- data.frame()
sig_files <- list.files(paste0(diffbind, "/result"), pattern="sig.tsv$", full.names=T, recursive=T)
comp_names <- gsub(".sig.tsv$", "", basename(sig_files))
  for (i in 1:length(sig_files)){
    sig <- read.table(sig_files[i], header=T, sep="\t")
    up <- length(which(sig$Fold > 0))
    dn <- length(which(sig$Fold < 0))
    up_dn <- data.frame(Comparison=comp_names[i], Up=up, Down=dn)
    sig_table <- rbind(sig_table, up_dn)
  }
print(kable(sig_table, caption=tabRef("peakSummary", "Differential peaks based on FDR < 0.05 and fold change > 1.5.")) %>%
  kable_styling() %>%
  htmltools::HTML())
```
<br>

```{r structure_link, echo=FALSE, results='asis', message=FALSE, warning=FALSE, eval=FALSE}
structureIndex<-ifelse(has_diffbind, 6, 5)
cat(paste0("## ", structureIndex, " Results folder structure\n"))
#cat(paste0("## Results folder structure\n"))

files_list <- list.dirs('./ChIP_seq_results', full.names=TRUE)

df<-NULL
peakFiles <- grep("macs2", files_list, value=T)
if(length(peakFiles) > 0){
  df<-rbind(df, data.frame(File=addLinkTag(peakFiles, peakFiles), Description="Peak file for each sample"))
}

homerFiles <- grep("homer_annotation", files_list, value=T)
if(length(homerFiles) > 0){
  df<-rbind(df, data.frame(File=addLinkTag(homerFiles, homerFiles), Description="Homer annotation and motif analysis"))
}

#bindiffFiles <- grep("diffbind", files_list, value=T)
#if(length(bindiffFiles) > 0){
#  df<-rbind(df, data.frame(File=addLinkTag(bindiffFiles, bindiffFiles), Description="Differential binding analysis"))
#}

print(kable(df, caption=tabRef("resultFiles", "Result files")) %>%
    kable_styling() %>%
    htmltools::HTML())

``` 

# Software versions
```{r softwareVersion, echo=FALSE, results='asis'}
versions<-display_versions()
```