---
title: "Cut&Run Analysis Report"
date: "`r format(Sys.time())`"
output: 
  rmdformats::readthedown:
    toc_depth: 3
    number_sections: true
---

```{css, echo=FALSE}
#content{
    max-width:1920px;
}
```

<style type="text/css">
    div.datatables { height: auto !important;}
</style>

```{r setup, include=FALSE}
library(XML)
library(rlist)
library(stringr)
library(kableExtra)
options(bitmapType='cairo')

knitr::opts_chunk$set(
  echo=FALSE,  
  message=FALSE, 
  warning=FALSE, 
  results = 'asis')

source('reportFunctions.R')
source('Pipeline.R')

myoptions_tbl<-read.table("fileList1.txt", header=FALSE, as.is=TRUE, fill=TRUE, sep="\t") |>
  dplyr::rename("Value"=V1, "Variable"=V2) |>
  dplyr::select(Variable, Value)

myoptions<-split(myoptions_tbl$Value, myoptions_tbl$Variable)
task_name = myoptions$task_name
cutrun_type = myoptions$cutrun_type
frag_120bp = myoptions$frag_120bp

print_table(myoptions_tbl, round_value=0, byDT=FALSE, row.names=FALSE)

task_df<-read.table("fileList2.txt", sep="\t", header=FALSE, as.is=TRUE)
task_df<-split(task_df$V1, task_df$V2)
```

```{r}
has_fastqc_raw="fastqc_raw_summary" %in% names(task_df)
has_fastqc_trimmed="trimmomatic_fastqc_summary" %in% names(task_df)
has_bowtie2_summary="bowtie2_summary" %in% names(task_df)

has_macs2_narrow_count="macs2_narrow_count" %in% names(task_df)
has_macs2_narrow_homer_motif="has_macs2_narrow_homer_motif" %in% names(task_df)

has_macs2_broad_count="macs2_broad_count" %in% names(task_df)
has_macs2_broad_homer_motif="macs2_broad_homer_motif" %in% names(task_df)

if ("homer_02_findPeaks_count" %in% names(task_df)){
  task_df$homer_count = task_df$homer_02_findPeaks_count
}else if ("homer_03_mergePeaks_count" %in% names(task_df)){
  task_df$homer_count = task_df$homer_03_mergePeaks_count
}
has_homer_count="homer_count" %in% names(task_df)

if("homer_03_motif" %in% names(task_df)){
  task_df$homer_motif = task_df$homer_03_motif
}else if("homer_04_motif" %in% names(task_df)){
  task_df$homer_motif = task_df$homer_04_motif
}
has_homer_motif="homer_motif" %in% names(task_df)

has_bamplot="bamplot" %in% names(task_df)
```
 
```{r,echo=FALSE,results='asis', eval = has_fastqc_raw}
cat("\n\n# Sequencing Quality\n\n")
prefix = paste0(task_df$fastqc_raw_summary, "/result/", task_name)
cat("\n\n## Base Quality\n\n")
include_graphics(paste0(prefix, ".FastQC.baseQuality.tsv.png"))
cat("\n\n## Sequence QC\n\n")
include_graphics(paste0(prefix, ".FastQC.sequenceGC.tsv.png"))
cat("\n\n## Read Count\n\n")
include_graphics(paste0(prefix, ".FastQC.reads.tsv.png"))
cat("\n\n## Adapter in Raw Data\n\n")
include_graphics(paste0(prefix, ".FastQC.adapter.tsv.png"))
```
  
```{r,echo=FALSE,results='asis', eval = has_fastqc_trimmed}
prefix = paste0(task_df$trimmomatic_fastqc_summary, "/result/", task_name)
cat("\n\n## Adapter after Trimming\n\n")
include_graphics(paste0(prefix, ".FastQC.adapter.tsv.png"))
```

```{r,echo=FALSE,results='asis', eval = has_bowtie2_summary}
cat("\n\n# Mapping Quality\n\n")
prefix = paste0(task_df$bowtie2_summary, "/result/", task_name)
cat("\n\n## Read in Chromosomes\n\n")
include_graphics(paste0(prefix, ".chromosome.png"))
cat("\n\n## Read in Samples\n\n")
include_graphics(paste0(prefix, ".reads.png"))

tlen_df<-read.table("fileList5.txt", sep="\t", header=FALSE, as.is=TRUE)
tlen_rmd="tlen.rmd"
tlen_content="\n\n# Template length\n\n"
tlen_content = paste0(tlen_content, "\n\ncutrun for ", cutrun_type, ", frag_120bp=", frag_120bp, "\n\n")

for (i in 1:nrow(tlen_df)){
  sample = tlen_df$V2[i]
  tlen = tlen_df$V1[i]
  tlen_content = paste0(tlen_content, "\n## ", sample, "\n\n")
  tlen_content = paste0(tlen_content, getFigure(tlen))
}
writeLines(tlen_content, tlen_rmd)
```

```{r child='tlen.rmd', eval=has_bowtie2_summary}
```

```{r}
add_peak_count<-function(peak_count, task_name, call_folder, call_name){
  count_file = paste0(call_folder, "/result/", task_name, ".txt")
  count_tbl = fread(count_file)
  colnames(count_tbl)[2] = call_name
  if(is.null(peak_count)){
    peak_count = count_tbl
  }else{
    peak_count = merge(peak_count, count_tbl, by="Sample", all=TRUE)
  }
  return(peak_count)
}

peak_count=NULL
if(has_macs2_narrow_count){
  peak_count = add_peak_count(peak_count, task_name, task_df$macs2_narrow_count, "macs2 narrow")
}

if(has_macs2_broad_count){
  peak_count = add_peak_count(peak_count, task_name, task_df$macs2_broad_count, "macs2 broad")
}

if(has_homer_count){
  peak_count = add_peak_count(peak_count, task_name, task_df$homer_count, "homer callPeaks")
}

if(!is.null(peak_count)){
  cat("\n\n# Peak Counts\n\n")
  print_table(peak_count, round_value=0, byDT=FALSE, row.names=FALSE)
}
```

```{r,echo=FALSE,results='asis', eval = has_bamplot}
rmd_file="bamplot.rmd"
rmd_content="\n\n# BAM plots\n\n"

prefix = paste0(task_df$bamplot, "/result/", task_name)
gff = fread(paste0(prefix, ".gff"))
genes = gff$V2
for(gene in genes){
  rmd_content = paste0(rmd_content, "\n## ", gene, "\n\n")
  rmd_content = paste0(rmd_content, getFigure(paste0(prefix, "_", gene, ".png")), "\n\n")
}
writeLines(rmd_content, rmd_file)
```

```{r child='bamplot.rmd', eval=has_bamplot}
```

```{r}
show_homer<-function(task_df, homer_task, homer_name) {
  homer_folder = task_df[[homer_task]]

  cat("#", homer_name, " (Top 5 of each sample)\n")
  dirs <- paste0(list.dirs(paste0(homer_folder, "/result"), full.names=T, recursive=F), "/knownResults")

  res_table <- data.frame()
  i=1
  for (i in 1:length(dirs)){
    curdir = dirs[i]
    knownRes_file <- paste0(curdir, ".txt")

    if(!file.exists(knownRes_file)){
      cat("\n\n<mark>No known motifs file found in ", basename(dirname(curdir)), ".</mark>\n\n")
      next
    }
    res_txt <- fread(knownRes_file, header=T, nrows=5, data.table=F)
    sample=basename(dirname(curdir))
    res_txt$Sample=sample
    res_txt$Motif="Not significant"

    for (j in 1:nrow(res_txt)) {
      svgfile = paste0(curdir, "/known", j, ".logo.svg")
      pngfile = paste0(curdir, "/known", j, ".logo.png")
      if (file.exists(svgfile)){
        if(!file.exists(pngfile)){
          x <- paste0("convert -density 1200 -resize 200x200 ", svgfile, " ", pngfile) 
        #cat(x, "\n")
          system(x)
        }
        res_txt$Motif[j]=paste0("<img src= ", pngfile, " >")
      }
    }

    res_table <- rbind(res_table, res_txt[,c("Motif Name", "P-value", "q-value (Benjamini)", "Sample", "Motif")])
  }

  tab <- tibble::tibble(Sample = res_table$Sample,
                        Name = gsub("/Homer","",res_table[,1]),
                        Motif = noquote(res_table$Motif),
                        Pvalue= res_table[,2], FDR = res_table[,3])
  print(kable(tab, caption= tabRef(homer_task, "Homer annotation for peaks"), escape = F) %>%
      kable_styling() %>%
      htmltools::HTML())
  #cat("\n\n**The entries without motif logo are not significant.**\n\n")
}
```

```{r,echo=FALSE,results='asis', eval = has_macs2_broad_homer_motif}
homer_task = "macs2_broad_homer_motif"
homer_name = "MACS2 Broad Peaks - Homer Motif Enrichment"
show_homer(task_df, homer_task, homer_name)
```

```{r,echo=FALSE,results='asis', eval = has_macs2_narrow_homer_motif}
homer_task = "macs2_narrow_homer_motif"
homer_name = "MACS2 Narrow Peaks - Homer Motif Enrichment"
show_homer(task_df, homer_task, homer_name)
```

```{r,echo=FALSE,results='asis', eval = has_homer_motif}
homer_task = "homer_motif"
homer_name = "Homer Peaks - Homer Motif Enrichment"
show_homer(task_df, homer_task, homer_name)
``` 
