---
title: "RNA Sequencing Data Analysis Report"
date: "`r format(Sys.time())`"
output: 
  html_document:
    toc: true
    toc_depth: 4
    number_sections: true
---

<style type="text/css">
.main-container {
  max-width: 90%;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r, child="Functions.Rmd"} 
```
  
```{r setup, include=FALSE}
library(kableExtra)

knitr::opts_chunk$set(fig.width=7, fig.height=6, out.width=800, out.height=600)

files<-read.table("fileList1.txt", header=FALSE, as.is=TRUE)
rownames(files)<-files$V2

rnaseqOptions<-read.table("fileList2.txt", header=FALSE, as.is=TRUE)
rownames(rnaseqOptions)<-rnaseqOptions$V2

hasFastqc<-any(grepl("fastqc_", rownames(files)))
hasFastqcPosttrim<-any(grepl("fastqc_post_trim_", rownames(files)))

hasSTAR<-any(grepl("STAR_", rownames(files)))
hasFeatureCount<-any(grepl("featureCounts_", rownames(files)))
hasDEseq2<-any(grepl("deseq2_", rownames(files)))
hasWebGestaltAnalysis<-any(grepl("WebGestalt_", rownames(files)))
hasGseaAnalysis<-any(grepl("gsea_", rownames(files)))
hasEnrichmentAnalysis<-hasWebGestaltAnalysis | hasGseaAnalysis

corNames<-gsub("_correlation_density", "", rownames(files)[grepl("_correlation_density", rownames(files))])
corNames<-sort(corNames)

getFigure<-function(category, description){
  return(paste0("```{r,echo=FALSE,results='asis', fig.align='center', fig.cap=figRef('", category, "', '",description, "', trunk.eval=file.exists(files['", category, "',1]))}\n",
"  check_and_include_graphics(files['", category, "',1])\n```\n"))
}

corrFile<-"corr.Rmd"
figureRmd<-function(corNames, section){
  result<-""
  for(i in c(1:length(corNames))){
    name=corNames[i]
    heatmap<-paste0(name, "_correlation_heatmap")
    pca<-paste0(name, "_correlation_PCA")
    corr_heatmap<-paste0(name, "_correlation_cluster")
    lines<-paste0("\n\n", section, " ", name, " samples\n\n", getFigure(heatmap, heatmap), getFigure(pca, pca), getFigure(corr_heatmap, corr_heatmap) )
    result<-paste0(result, lines)
  }
  return(result)
}
cat(figureRmd(corNames, "###"), file=corrFile)

```

<br>
  
# Sequencing Quality
  
## Summary of sequencing quality
  
```{r fastqc_raw_per_base_sequence_quality,echo=FALSE,results='asis', include = hasFastqc, eval = hasFastqc, fig.align="center", fig.cap=figRef("fastqc_raw_per_base_sequence_quality", "The mean quality scores of sequencing reads in each position", trunk.eval=file.exists(files["fastqc_raw_per_base_sequence_quality",1]))}
check_and_include_graphics(files["fastqc_raw_per_base_sequence_quality",1])
```

<br>

```{r fastqc_raw_per_sequence_gc_content,echo=FALSE,results='asis', include = hasFastqc, eval = hasFastqc, fig.align="center", fig.cap=figRef("fastqc_raw_per_sequence_gc_content", "The average GC content of sequencing reads", trunk.eval=file.exists(files["fastqc_raw_per_sequence_gc_content",1]))}
check_and_include_graphics(files["fastqc_raw_per_sequence_gc_content",1])
```

<br>
  
```{r fastqc_raw_adapter_content,echo=FALSE,results='asis', include = hasFastqc, eval = hasFastqc, fig.align="center", fig.cap=figRef("fastqc_raw_adapter_content", "Adapter content of sequencing reads", trunk.eval=file.exists(files["fastqc_raw_adapter_content",1]))}
check_and_include_graphics(files["fastqc_raw_adapter_content",1])
```

<br>
  
```{r fastqc_post_trim_adapter_content,echo=FALSE,results='asis', include = hasFastqc, eval = hasFastqc, fig.align="center", fig.cap=figRef("fastqc_post_trim_adapter_content", "Adapter content of sequencing reads after cutadapt", trunk.eval=file.exists(files["fastqc_post_trim_adapter_content",1]))}
check_and_include_graphics(files["fastqc_post_trim_adapter_content",1])
```

<br>
  
# Mapping quality
## Summary of mapping quality
```{r STAR_summary,echo=FALSE,results='asis', include = hasSTAR, eval = hasSTAR, fig.align="center", fig.cap=figRef("STAR_summary", "The statistics of RNAseq mapping results")}
include_graphics(files["STAR_summary",1])
```

<br>

```{r STAR_summary_table, echo=FALSE, results='asis', include = hasSTAR, eval = hasSTAR}
table<-read.csv(files["STAR_summary_table",1])
output_table1<-data.frame(table[,c(1,2,3,4,5)], table[,3]/table[,2], (table[,3]+table[,4])/table[,2])
colnames(output_table1)<-c("Sample", "Input reads", "Uniquely mapped reads", "Mapped to multiple loci", "Mapped to too many loci", "Uniquely mapped ratio", "Mapped ratio")

box_stats<-boxplot(output_table1$`Uniquely mapped ratio`, plot=FALSE)$stats
print(kable_styling(kable(output_table1, caption=tabRef("mapSummary", "The summary of RNAseq mapping results"), row.names=F) %>% 
          row_spec(which(output_table1$`Uniquely mapped ratio` > box_stats[5]), color = "black", background = "bisque") %>% 
          row_spec(which(output_table1$`Uniquely mapped ratio` < box_stats[1]), color = "black", background = "honeydew")))
cat("**Uniquely mapped ratio** = Uniquely mapped / Input reads\n\n**Mapped ratio** = (Uniquely mapped + Mapped to multiple loci)/ Input reads\n\n")
```

<br>

# Expression Quantification
## Summary of reads assignment

```{r, results="asis", echo=FALSE, include = hasFeatureCount, eval = hasFeatureCount} 
cat("featureCounts was used to count reads mapped genes, and **MULTI** mapped reads were ")
if(rnaseqOptions["featureCounts_UseMultiMappingReads", 1] == "FALSE"){
  cat(" **NOT USED**. \n\n")
}else{
  cat(" **USED**. \n\n")
}
```

```{r,echo=FALSE,results='asis', include = hasFeatureCount, eval = hasFeatureCount, fig.align="center", fig.cap=figRef("gene_reads", "The number of reads assigned to genes in uniquely mapped reads")}
# For featureCounts, unassigned_unmapped -> 
#                    unassigned_mappingquality -> 
#                    unassigned_fragmentlength -> 
#                    unassigned_chimericreads -> 
#                    unassigned_duplicate -> 
#                    unassigned_multimapping ->
#                    unassigned_secondary ->
#                    unassigned_junction_condition ->
#                    assigned
# So, we cannot get how many unique mapped reads from the featureCounts summary.
include_graphics(files["featureCounts_table_png",1])
```

<br>

```{r echo=FALSE,results='asis', include = hasFeatureCount, eval = hasFeatureCount}
table<-read.csv(files["featureCounts_table",1])
readsUsed = "reads"
if(rnaseqOptions["featureCounts_UseMultiMappingReads", 1] == "FALSE"){
  readsUsed = "**UNIQUE** reads"
}
colnames(table)[1]<-"Sample"
box_stats<-boxplot(table$Percent_Assigned, plot=FALSE)$stats
print(kable_styling(kable(table, caption=tabRef("geneSummary", paste0("The summary of ", readsUsed, " assignment to genes")), row.names=F)%>% 
          row_spec(which(table$Percent_Assigned > box_stats[5]), color = "black", background = "bisque") %>% 
          row_spec(which(table$Percent_Assigned < box_stats[1]), color = "black", background = "honeydew")))
```

<br>

## Expression density of samples
```{r correlation_density,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("correlation_density", "Expression density distribution of each sample", trunk.eval=file.exists(files["all_correlation_density",1]))}
check_and_include_graphics(files["all_correlation_density",1])
```

<br>

## Similarity between samples
```{r, child="corr.Rmd"} 
```

<br>

```{r deseq2_1, results="asis", echo=FALSE} 
if(hasDEseq2 ){
  cat("# Differential expression\n<br>\n\n")
  if ("DE_use_raw_pvalue" %in% row.names(rnaseqOptions) & rnaseqOptions["DE_use_raw_pvalue", 1]==0) {pValueText="FDR"} else {pValueText="p value"}
  cat("Differential expression analysis criteria:  fold change >= ", rnaseqOptions["DE_fold_change", 1], " and ",pValueText," <= ", rnaseqOptions["DE_pvalue", 1], "\n\n")
  cat("## Volcano plot\n<br>\n\n")
}
```

```{r deseq2_1_volcano_plot,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("Volcano_plot", "Volcano plot for all comparisons", trunk.eval=file.exists(files["deseq2_volcano_plot",1]))}
check_and_include_graphics(files["deseq2_volcano_plot",1])
```

<br>

```{r deseq2_2, results="asis", echo=FALSE} 
if(hasDEseq2 ){
  cat("## Summary of differential analysis\n<br>\n")
}
```


```{r deseq2_2_sig, echo=FALSE,results='asis'}
if(hasDEseq2 ){
  DESeq2_files<-files[grep("DESeq2_sig", files[,1]),]
  DESeq2_list<-list()
  for (i in 1:nrow(DESeq2_files)){
    DESeq2_list[[i]]<-read.csv(DESeq2_files[i,1], header=TRUE)
  }
  output_table3<-data.frame(c(gsub("_DESeq2_sig.csv","",basename(DESeq2_files[,1]))), c(unlist(lapply(DESeq2_list, nrow))))
  colnames(output_table3)<-c("Comparison", "Number of DEGs")
  print(kable_styling(kable(output_table3, caption= tabRef("deseq2_table", paste0("DEG numbers under FC > ", rnaseqOptions["DE_fold_change", 1], " & ",pValueText," < ", rnaseqOptions["DE_pvalue", 1])))))

  for (i in 1:nrow(DESeq2_files)){
    comparisonName<-DESeq2_files[i,2]
    deseq2<-read.csv(DESeq2_files[i,1], header=TRUE, row.names=1)
    if(nrow(deseq2) > 0){
      maxrow<-min(nrow(deseq2), 10)
      cnames<-c("baseMean", "log2FoldChange","lfcSE", "stat", "pvalue", "padj", "FoldChange")
      if("Feature_gene_name"%in%colnames(deseq2)){
        cnames<-c("Feature_gene_name", cnames)
      }
      deseq2table<-deseq2[c(1:maxrow), cnames]
      print(kable_styling(kable(deseq2table, row.names=1, caption= tabRef(comparisonName, paste0(comparisonName, " top ", maxrow, " differential expressed genes")))%>% 
          row_spec(which(deseq2table$FoldChange > 1), color = "black", background = "bisque") %>% 
          row_spec(which(deseq2table$FoldChange < 1), color = "black", background = "honeydew")))
    }
  }
}
```

<br>

```{r enrich, results="asis", echo=FALSE} 
if(hasEnrichmentAnalysis ){
  cat("# Funcional enrichment analysis\nList of the top five significantly enriched elements\n<br>\n")
}
```

<br>
  
```{r webgestalt, results="asis", echo=FALSE} 
if(hasWebGestaltAnalysis){
  cat(paste0("  \n## WebGestalt  \n"))
  enrich_files<-files[grepl("WebGestalt_", rownames(files)),]
  enrich_files$Comparisons<-gsub("WebGestalt_GO....", "", enrich_files$V2)
  enrich_files$Comparisons<-gsub("WebGestalt_KEGG_", "", enrich_files$Comparisons)
  comparisons<-unique(enrich_files$Comparisons)
  for (j in 1:length(comparisons)){
    comparison<-comparisons[j]
    comp_files<-enrich_files[enrich_files$Comparisons == comparison,]
    cat(paste0("  \n### ", comparison, "  \n"))
  
    for (i in 1:nrow(comp_files)){
      if (!file.exists(comp_files[i,1])) {
        next;
      }
      ename<-rownames(comp_files)[i]
      if(grepl(".html.rds", comp_files[i,1])){
        plotData <-readRDS(comp_files[i,1])
        fdata<-plotData$enriched
        fdata<-fdata[c(1:min(nrow(fdata),5)),c(1,2,4,5,6,7,8,9,12,13)]
        print(kable_styling(kable(fdata, caption=tabRef(ename, ename))) %>% 
          row_spec(which(fdata$geneUp > fdata$geneDown), color = "black", background = "bisque") %>% 
          row_spec(which(fdata$geneUp < fdata$geneDown), color = "black", background = "honeydew"))
        cat("\n\n<hr>")
      }else{
        #txtFile<-gsub(".html.rds$", "", comp_files[i,1])
        #fdata<-read.table(txtFile, sep="\t", header=T, stringsAsFactors = F)
        fdata<-read.table(comp_files[i,1], sep="\t", header=T, stringsAsFactors = F)
        fdata<-fdata[c(1:min(nrow(fdata))),]
        fdata$geneSet<-paste0("[", fdata$geneSet, "](", fdata$link, ")")
        fdata<-fdata[,c(1,2,4,5,6,7,8,9)]
        print(kable_styling(kable(fdata, caption=tabRef(ename, ename))))
      }
    }  
  }
}
```

<br>
  
```{r gsea, results="asis", echo=FALSE, warning=FALSE} 
if(hasGseaAnalysis){
  cat(paste0("  \n## GSEA  \n"))
  processGesaTable<-function(gseaTableFile,maxCategoryFdr=0.05,maxCategory=5,absLinkPath=FALSE) {
    rawTable<-read.delim(gseaTableFile,header=T,as.is=T)
    rawTableOut<-NULL
    for (i in head(which(rawTable$FDR.q.val<=maxCategoryFdr),maxCategory)) {
      categoryName=rawTable[i,1]
      gseaUrl=paste0("http://software.broadinstitute.org/gsea/msigdb/cards/",categoryName)
      categoryNameInTable<-addLinkTag(text=categoryName,link=gseaUrl)
      
      gseaWeb<-getURL(gseaUrl)
      temp<-strsplit(gseaWeb,"description|\\<td\\>")[[1]]
      j=grep("Brief",temp)
      categoryDescription<-gsub("^>","",temp[j+2])
      categoryDescription<-gsub("<\\/$","",categoryDescription)
      
      rawTableOut<-rbind(rawTableOut,unlist(c(categoryNameInTable,categoryDescription,rawTable[i,c(4,5,6,8)])))
    }
    if(!is.null(rawTableOut)){
      colnames(rawTableOut)[1:2]<-c("Name","Description")
    }
    return(rawTableOut)
  }
  
  gsea_files<-files[grepl("gsea_", rownames(files)),]
  gsea_files$Comparisons<-gsub("gsea_", "", gsea_files$V2)
  comparisons<-unique(gsea_files$Comparisons)
  for (j in 1:length(comparisons)){
    comparison<-comparisons[j]
    comp_files<-gsea_files[gsea_files$Comparisons == comparison,]
    cat(paste0("  \n### ", comparison, "  \n"))
  
    for (i in 1:nrow(comp_files)){
      gfolders<-read.csv(comp_files[i,1],header=T,stringsAsFactor=F)
      gname<-gsub(".rnk.*", "", basename(comp_files[i,1]))
      for (j in 1:nrow(gfolders)){
        resultDirSub<-gfolders$Folder[j]
        
        posTableFile<-list.files(resultDirSub,pattern="gsea_report_for_na_pos_\\d+\\.xls$",full.names=TRUE)
        if (length(posTableFile)!=1) {
          warning(paste0("Can't find positive-regulated GSEA table file in ", resultDirSub))
        }else{
          rawTableOut<-processGesaTable(posTableFile)
          if(!is.null(rawTableOut)){
            ename<-paste0(gname, " ", gfolders$GseaCategory[j], " Positive-regulated")
            print(kable_styling(kable(rawTableOut, caption=tabRef(ename, ename))))
          }else{
            warning(paste0("Can't find significant positive-regulated GSEA gene set in ", resultDirSub))
          }
        }
        
        negTableFile<-list.files(resultDirSub,pattern="gsea_report_for_na_neg_\\d+\\.xls$",full.names=TRUE)
        if (length(negTableFile)!=1) {
          warning(paste0("Can't find negative-regulated GSEA table file in ", resultDirSub))
        }else{
          rawTableOut<-processGesaTable(negTableFile)
          if(!is.null(rawTableOut)){
            ename<-paste0(gname, " ", gfolders$GseaCategory[j], " Negative-regulated")
            print(kable_styling(kable(rawTableOut, caption=tabRef(ename, ename))))
          }else{
            warning(paste0("Can't find significant negative-regulated GSEA gene set in ", resultDirSub))
          }
        }
      }
    }  
  }
}
``` 

<br>
  
# Results folder structure

```{r structure_link, echo=FALSE, results='asis'}

df<-NULL
countFile<-list.files('.',pattern=".count$",full.names=TRUE,recursive=TRUE)
if(length(countFile) > 0){
  df<-rbind(df, data.frame(File=addLinkTag(countFile, countFile), Description="Gene expression read count table"))
}

fpkmFile<-list.files('.',pattern="fpkm.tsv$",full.names=TRUE,recursive=TRUE)
if(length(fpkmFile) > 0){
  df<-rbind(df, data.frame(File=addLinkTag(fpkmFile, fpkmFile), Description="Gene expression abundance table"))
}

deseqAll<-list.files('.',pattern="_DESeq2.csv$",full.names=TRUE,recursive=TRUE)
if(length(deseqAll) > 0){
  df<-rbind(df, data.frame(File=addLinkTag(deseqAll, deseqAll), Description="Differential expression analysis table"))
}

deseqSig<-list.files('.',pattern="_DESeq2_sig.csv$",full.names=TRUE,recursive=TRUE)
if(length(deseqSig) > 0){
  df<-rbind(df, data.frame(File=addLinkTag(deseqSig, deseqSig), Description="Significantly differential expressed genes"))
}

webgestaltDeseq2<-list.files('.',pattern=".txt.html$",full.names=TRUE,recursive=TRUE)
if(length(webgestaltDeseq2) > 0){
  df<-rbind(df, data.frame(File=addLinkTag(webgestaltDeseq2, webgestaltDeseq2), Description="WebgestaltR enrichment table"))
}

print(kable_styling(kable(df, caption=tabRef("resultFiles", "Result files"))))

```

# Software versions
```{r softwareVersion, echo=FALSE, results='asis'}
files<-read.delim("fileList4.txt", header=F, stringsAsFactors = F)
df<-NULL
curfile<-files$V1[1]
for(curfile in files$V1){
  if(file.exists(curfile)){
    curdf<-read.csv(curfile, header=F)
    df<-rbind(df, curdf)
  }
}

if(file.exists("fileList5.txt")){
  vers<-read.delim("fileList5.txt", header=F, stringsAsFactors = F)
  vers<-vers[,c(2,1)]
  colnames(vers)<-c("V1","V2")
  df<-rbind(df, vers)
}

df<-unique(df)
df<-df[order(df$V1),]
colnames(df)<-c("Software", "Version")
print(kable_styling(kable(df, caption=tabRef("versionFiles", "Software versions"))))

```
