---
title: "smallRNA Sequencing Analysis Report"
date: "`r format(Sys.time())`"
output: 
  html_document:
    toc: true
    toc_depth: 2
    number_sections: true
---

<style type="text/css">
.main-container {
  max-width: 90%;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r, child="Functions.Rmd"} 
```
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=7, fig.height=6, out.width=800, out.height=600)

files<-read.table("fileList1.txt", header=FALSE, as.is=TRUE)
rownames(files)<-files$V2

#For RnaSeq only
#rnaseqOptions<-read.table("fileList2.txt", header=FALSE, as.is=TRUE)
#rownames(rnaseqOptions)<-rnaseqOptions$V2

check_and_include_graphics<-function(graphicFile) {
  if (file.exists(graphicFile)) {
    include_graphics(graphicFile)
  }
}

getCategory<-function(category, description){
  return(paste0("```{r,echo=FALSE,results='asis', fig.align='center', fig.cap=figRef('", category, "', '",description, "')}\n",
"  check_and_include_graphics(files['", category, "',1])\n```\n"))
}

getCategory2<-function(category1, category2, description){
  return(paste0("```{r,echo=FALSE,results='asis', fig.show='hold', out.width='45%', fig.align = 'default'}\n",
"check_and_include_graphics(files['", category1, "',1])\n",
"check_and_include_graphics(files['", category2, "',1])\n",
"```\n",
"```{r,echo=FALSE,results='asis'}\n",
"curFigRef<-figRef('", paste0(category1, category2), "', '",description, "')\n",
"```\n",
"<center>`r curFigRef`</center>\n"))
}

categoryRmd<-function(dfCategory, fileName, header){
  hasFile<-FALSE
  result<-""
  for(i in c(1:nrow(dfCategory))){
    section=dfCategory$Section[i]
    category=dfCategory$Category[i]
    name=dfCategory$Name[i]
    heatmap<-paste0("correlation_", category, "_heatmap")
    pca<-paste0("correlation_", category, "_pca")
    group_heatmap<-paste0("correlation_", category, "_group_heatmap")
    corr_heatmap<-paste0("correlation_", category, "_corr_cluster")
    
    if(file.exists(files[heatmap,1])){
      hasFile<-TRUE
    }
    
    lines<-paste0(section, " ", name, "\n", getCategory2(heatmap, pca, paste0("Heatmap (left) and PCA (right) of ", name) ), getCategory2(group_heatmap, corr_heatmap, paste0("Group heatmap (left) and group correlation (right) of ", name)) )
    result<-paste0(result, lines)
  }

  if(hasFile){
    cat(paste0(header, result), file=fileName)
    return(TRUE)
  }else{
    cat("", file=fileName)
    return(FALSE)
  }
}

hasLengthDistCategory<-file.exists(files["host_length_dist_category",1])

hasAdapterRaw<-file.exists(files["fastqc_raw_adapter_content",1])
hasAdapterPostTrim<-file.exists(files["fastqc_post_trim_adapter_content",1])
hasAdapterRawOnly<-hasAdapterRaw & (!hasAdapterPostTrim)

hasFastqc<-any(grepl("fastqc_", rownames(files)))
hasFastqcPosttrim<-any(grepl("fastqc_post_trim_", rownames(files)))

hostGenomeFile="hostgenome.Rmd"
hostGenome<-data.frame(Section=c("##", "##"), Category=c("mirna", "trna"), Name=c("miRNA", "tRNA"))
hasHostGenome<-categoryRmd(hostGenome, hostGenomeFile, "")

nonhostGenomeFile="nonhostgenome.Rmd"
nonhostGenome<-data.frame(Section=c("###", "###"), Category=c("group1", "group2"), Name=c("Microbime bacteria", "Environment bacteria"))
hasNonhostGenome<-categoryRmd(nonhostGenome, nonhostGenomeFile, "## Non-host genome\n\n")

nonhostLibraryFile="nonhostlibrary.Rmd"
nonhostLibrary<-data.frame(Section=c("###", "###"), Category=c("trnalib", "rrnalib"), Name=c("tRNA library", "rRNA library"))
hasNonhostLibrary<-categoryRmd(nonhostLibrary, nonhostLibraryFile, "## Non-host library\n\n")

```

<br>

# Summary
This report includes summary (section 1), preprocessing (section 2), host genome (section 3)
```{r, echo=FALSE, results='asis'}
if(hasNonhostGenome){
  cat(", nonhost genome (section 4)")
  nextIndex = 5
}else{
  nextIndex = 4
}

if(hasNonhostLibrary){
  cat(", nonhost library (section ", nextIndex, ")")
}

```

```{r,echo=FALSE,results='asis', fig.show="hold", out.width="45%", fig.align = "default", out.height=1024}
cat("## Read summary count\n")
check_and_include_graphics(files["read_summary_count",1])
check_and_include_graphics(files["read_summary_perc",1])
```

```{r,echo=FALSE,results='asis'}
fig1ref<-figRef("read_summary", "Overall read count (left) and read percentage (right) distribution")
```
<center>`r fig1ref`</center>

<br>

# Preprocessing
  
```{r fastqc_raw_adapter_content,echo=FALSE,results='asis', include=hasAdapterRawOnly, eval=hasAdapterRawOnly, fig.align="center", fig.cap=figRef("fastqc_adapter", "Adapter content of sequencing reads", trunk.eval=hasAdapterRawOnly)}
cat("## Adapter\n")
check_and_include_graphics(files["fastqc_raw_adapter_content",1])
```

```{r,echo=FALSE,results='asis', fig.show="hold", out.width="45%", fig.align = "default", include=hasAdapterPostTrim, eval=hasAdapterPostTrim}
cat("## Adapter before and after trimming\n")
check_and_include_graphics(files["fastqc_raw_adapter_content",1])
check_and_include_graphics(files["fastqc_post_trim_adapter_content",1])
```

```{r,echo=FALSE,results='asis', include=hasAdapterPostTrim, eval=hasAdapterPostTrim}
fig1ref<-figRef("fastqc_adapter_content", "Raw adapter (left) and post-trim adapter (right)")
cat(paste0("<center>", fig1ref, "</center>\n"))
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("fastq_len", "Read length distribution after adapter trimming", trunk.eval=(file.exists(files["fastq_len",1]) & (!hasLengthDistCategory)))}
cat("## Read length distribution after adapter trimming\n")
check_and_include_graphics(files["fastq_len",1])
```

# Host genome
## Overall 
```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("category_mapped_bar", "Read distribution", trunk.eval=file.exists(files["category_mapped_bar",1]))}
check_and_include_graphics(files["category_mapped_bar",1])
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("category_mapped_group", "Read distribution in grouped samples", trunk.eval=file.exists(files["category_mapped_group",1]))}
check_and_include_graphics(files["category_mapped_group",1])
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("category_smallrna_bar", "smallRNA read distribution", trunk.eval=file.exists(files["category_smallrna_bar",1]))}
check_and_include_graphics(files["category_smallrna_bar",1])
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("category_smallrna_group", "smallRNA read distribution in grouped samples", trunk.eval=file.exists(files["category_smallrna_group",1]))}
check_and_include_graphics(files["category_smallrna_group",1])
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("host_length_dist_category", "Host read length distribution in categories", trunk.eval=file.exists(files["host_length_dist_category",1]))}
check_and_include_graphics(files["host_length_dist_category",1])
```

<br>

```{r hostgenome, child=hostGenomeFile} 
```

```{r,echo=FALSE,results='asis', fig.align="center", eval=file.exists(files["deseq2_host_vis",1]), fig.cap=figRef("deseq2_host_vis", "DESeq2 result of host smallRNA", trunk.eval=file.exists(files["deseq2_host_vis",1]))}
cat('## host smallRNA DESeq2 result\n') 

check_and_include_graphics(files["deseq2_host_vis",1])
```

```{r, echo=FALSE, results='asis'}
if(hasNonhostGenome | hasNonhostLibrary){
  cat("# Non-Host")
}

```
<br>

```{r nonhostgenome, include=hasNonhostGenome, child=nonhostGenomeFile} 
```

<br>

```{r nonhostlibrary, include=hasNonhostLibrary, child=nonhostLibraryFile} 
```

```{r teardown, include=FALSE}
if(file.exists(hostGenomeFile)){
  file.remove(hostGenomeFile)
}
if(file.exists(nonhostGenomeFile)){
  file.remove(nonhostGenomeFile)
}
if(file.exists(nonhostLibraryFile)){
  file.remove(nonhostLibraryFile)
}
```

```{r, echo=FALSE, results='asis'}
if(hasNonhostGenome | hasNonhostLibrary){
  cat("## Non-host summary")
}
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("nonhost_overlap_bar", "Read overlapped in non-host genome/database", trunk.eval=file.exists(files["nonhost_overlap_bar",1]))}
check_and_include_graphics(files["nonhost_overlap_bar",1])
```
