---
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
output: 
  rmdformats::readthedown:
    toc_depth: 3
    number_sections: true
---

```{css, echo=FALSE}
#content{
    max-width:1920px;
}
```

```{r setup, include=FALSE}
library(kableExtra)

source("Functions.R")
source("Pipeline.R")

knitr::opts_chunk$set(echo = FALSE,  message=FALSE, warning=FALSE, results = 'hide', fig.width=8, fig.height = 6,tidy = TRUE, tidy.opts = list(comment = FALSE))

files<-read.table("fileList1.txt", header=FALSE, as.is=TRUE)
files=split(files$V1, files$V2)

myoptions<-fread("fileList2.txt", header=FALSE)
myoptions$V1 <- as.character(myoptions$V1)
option_map<-split(myoptions$V1, myoptions$V2)

#For RnaSeq only
#rnaseqOptions<-read.table("fileList2.txt", header=FALSE, as.is=TRUE)
#rownames(rnaseqOptions)<-rnaseqOptions$V2

has_file<-function(files, name){
  if(name %in% names(files)){
    curfile=files[[name]]
    return(file.exists(curfile))
  }else{
    return(FALSE);
  }
}

get_file<-function(files, name){
  if(name %in% names(files)){
    return(files[[name]])
  }else{
    return("");
  }
}

check_and_include_graphics<-function(graphicFile) {
  if (file.exists(graphicFile)) {
    include_graphics(graphicFile)
  }
}

getCategory<-function(category, description){
  return(paste0("```{r,echo=FALSE,results='asis', fig.align='center', fig.cap=figRef('", category, "', '",description, "')}\n",
"  check_and_include_graphics(get_file(files,'", category, "'))\n```\n"))
}

getCategory2<-function(category1, category2, description){
  return(paste0("```{r,echo=FALSE,results='asis', fig.show='hold', out.width='45%', fig.align = 'default'}\n",
"has_category_file<-has_file(files,'", category1, "')\n",
"check_and_include_graphics(get_file(files,'", category1, "'))\n",
"check_and_include_graphics(get_file(files,'", category2, "'))\n",
"```\n",
"```{r,echo=FALSE,results='asis',eval=has_category_file}\n",
"curFigRef<-figRef('", paste0(category1, category2), "', '",description, "')\n",
"cat(paste0('<center>', curFigRef, '</center>'))\n",
"```\n"))
}

categoryRmd<-function(dfCategory, fileName, header){
  hasFile<-FALSE
  result<-""
  i=1
  for(i in c(1:nrow(dfCategory))){
    section=dfCategory$Section[i]
    category=dfCategory$Category[i]
    name=dfCategory$Name[i]
    heatmap<-paste0("correlation_", category, "_heatmap")
    pca<-paste0("correlation_", category, "_pca")
    group_heatmap<-paste0("correlation_", category, "_group_heatmap")
    corr_heatmap<-paste0("correlation_", category, "_corr_cluster")
    barplot<-paste0(category, "_barplot")

    if(has_file(files, heatmap)){
      hasFile<-TRUE
    }
    
    lines<-paste0(section, " ", name, "\n", getCategory2(heatmap, pca, paste0("Heatmap (left) and PCA (right) of ", name) ), getCategory2(group_heatmap, corr_heatmap, paste0("Group heatmap (left) and group correlation (right) of ", name)) )
    result<-paste0(result, lines)

    if(file.exists('fileList7.txt')){
      tbl=read.table('fileList7.txt', sep="\t", header=FALSE, stringsAsFactor=FALSE)
      comparisons = unique(tbl[,ncol(tbl)])
      comparison=comparisons[1]
      for(comparison in comparisons){
        volcano<-paste0("deseq2_", category, "_volcano_", comparison)
        result<-paste0(result, getCategory(volcano, paste0("Volcano plot of comparison ", comparison, "of ", name)))
      }
    }

    if(has_file(files, barplot)){
      result<-paste0(result, getCategory(barplot, paste0("Barplot of ", name)))
    }
  }

  if(hasFile){
    cat(paste0(header, result), file=fileName)
    return(TRUE)
  }else{
    cat("", file=fileName)
    return(FALSE)
  }
}

hasLengthDistCategory<-has_file(files,"host_length_dist_category")

has_fastq_len = has_file(files, "fastq_len") & (!hasLengthDistCategory)

hasAdapterRaw<-has_file(files,"fastqc_raw_adapter_content")
hasAdapterPostTrim<-has_file(files,"fastqc_post_trim_adapter_content")
hasAdapterRawOnly<-hasAdapterRaw & (!hasAdapterPostTrim)

has_category_mapped_bar<-has_file(files, "category_mapped_bar")

hasFastqc<-any(grepl("fastqc_", names(files)))
hasFastqcPosttrim<-any(grepl("fastqc_post_trim_", names(files)))

hostGenomeFile="hostgenome.Rmd"
hostGenome<-data.frame(Section=c("##"), Category=c("mirna"), Name=c("miRNA"))

hasTRNA<-any(grepl("correlation_trna_", names(files)))
if(hasTRNA){
  hostGenome<-rbind(hostGenome, data.frame(Section=c("##"), Category=c("trna"), Name=c("tRNA")))
}

hasRRNA<-any(grepl("correlation_rrna_", names(files)))
if(hasRRNA){
  hostGenome<-rbind(hostGenome, data.frame(Section=c("##"), Category=c("rrna"), Name=c("rRNA")))
}

hasYRNA<-any(grepl("correlation_yrna_", names(files)))
if(hasYRNA){
  hostGenome<-rbind(hostGenome, data.frame(Section=c("##"), Category=c("yrna"), Name=c("yRNA")))
}

hasSNRNA<-any(grepl("correlation_snrna_", names(files)))
if(hasSNRNA){
  hostGenome<-rbind(hostGenome, data.frame(Section=c("##"), Category=c("snrna"), Name=c("snRNA")))
}

hasSNORNA<-any(grepl("correlation_snorna_", names(files)))
if(hasSNORNA){
  hostGenome<-rbind(hostGenome, data.frame(Section=c("##"), Category=c("snorna"), Name=c("snoRNA")))
}

hasERV<-any(grepl("correlation_ERV_", names(files)))
if(hasERV){
  hostGenome<-rbind(hostGenome, data.frame(Section=c("##"), Category=c("ERV"), Name=c("ERV")))
}

hasHostGenome<-categoryRmd(hostGenome, hostGenomeFile, "")

nonhostGenomeFile="nonhostgenome.Rmd"
nonhostLibraryFile="nonhostlibrary.Rmd"
nonhostBacteriaAggFile="nonhostBacteriaAgg.Rmd"
nonhostBacteriaEstFile="nonhostBacteriaEst.Rmd"

if(file.exists('fileList8.txt')){
  nggs=read.table('fileList8.txt', sep="\t", header=FALSE, stringsAsFactor=FALSE)
  nonhostGenome<-data.frame(Section="###", Category=nggs$V2, Name=nggs$V1)
  hasNonhostGenome<-categoryRmd(nonhostGenome, nonhostGenomeFile, "## Non-host genome\n\n")

  nonhostLibrary<-data.frame(Section=c("###", "###"), Category=c("trnalib", "rrnalib"), Name=c("tRNA library", "rRNA library"))
  hasNonhostLibrary<-categoryRmd(nonhostLibrary, nonhostLibraryFile, "## Non-host library\n\n")

  categories = c("species", "genus", "family", "order", "class", "phylum")

  # nonhostBacteriaAgg=NULL
  # for(cat in categories){
  #   nonhostBacteriaAgg<-rbind(nonhostBacteriaAgg, data.frame(Section=c("###"), Category=c(paste0("refseq_bacteria_", cat, "_agg")), Name=c(paste0("Bacteria ", cat, " aggregated count"))))
  # }
  # hasRefBacteriaAgg<-categoryRmd(nonhostBacteriaAgg, nonhostBacteriaAggFile, "## Refseq bacteria aggregated count \n\n")
  hasRefBacteriaAgg<-FALSE

  nonhostBacteriaEst=NULL
  for(cat in categories){
    nonhostBacteriaEst<-rbind(nonhostBacteriaEst, data.frame(Section=c("###"), Category=c(paste0("refseq_bacteria_", cat, "_est")), Name=c(paste0("Bacteria ", cat, " estimated count"))))
  }
  hasRefBacteriaEst<-categoryRmd(nonhostBacteriaEst, nonhostBacteriaEstFile, "## Refseq bacteria estimated count \n\n")
}else{
  hasNonhostGenome<-FALSE
  hasNonhostLibrary<-FALSE
  hasRefBacteriaAgg<-FALSE
  hasRefBacteriaEst<-FALSE
}

hasHostDEseq2<-any(grepl("deseq2_host_vis", names(files)))
hasNonhostDEseq2<-any(grepl("deseq2_nonhost_vis", names(files)))

if ("DE_use_raw_pvalue" %in% names(option_map) & option_map$DE_use_raw_pvalue=="1") {
  pValueText="p value"
} else {
  pValueText="FDR adjusted p value"
}
deseq2criteria<-paste0("absolute fold change >= ", option_map$DE_fold_change, " and ", pValueText, " <= ", option_map$DE_pvalue)

#the gene count table doesn't make sense. We will not put it into final report right now.
has_custom_genome_gene<-has_file(files, "custom_group_gene_count")
```

---
title: SmallRNA Sequencing Analysis Report of Project `r gsub(":", " ", option_map['task_name'])`
author:
- name: `r option_map['email']`
  affiliation: `r option_map['affiliation']`
---


<br>

# Summary

This report includes summary (section 1), preprocessing (section 2), host genome (section 3)

```{r, echo=FALSE, results='asis'}
if(hasNonhostGenome){
  cat(", nonhost genome (section 4)")
  nextIndex = 5
}else{
  nextIndex = 4
}

if(hasNonhostLibrary){
  cat(", nonhost library (section ", nextIndex, ")")
}

```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("read_summary", "Read summary")}
cat("## Read summary count\n")
check_and_include_graphics(files$read_summary)
```

<br>

# Preprocessing
  
```{r fastqc_raw_adapter_content,echo=FALSE,results='asis', include=hasAdapterRawOnly, eval=hasAdapterRawOnly, fig.align="center", fig.cap=figRef("fastqc_adapter", "Adapter content of sequencing reads", trunk.eval=hasAdapterRawOnly)}
cat("## Adapter\n")
check_and_include_graphics(files$fastqc_raw_adapter_content)
```

```{r,echo=FALSE,results='asis', fig.show="hold", out.width="45%", fig.align = "default", include=hasAdapterPostTrim, eval=hasAdapterPostTrim}
cat("## Adapter before and after trimming\n")
check_and_include_graphics(files$fastqc_raw_adapter_content)
check_and_include_graphics(files$fastqc_post_trim_adapter_content)
```

```{r,echo=FALSE,results='asis', include=hasAdapterPostTrim, eval=hasAdapterPostTrim}
fig1ref<-figRef("fastqc_adapter_content", "Raw adapter (left) and post-trim adapter (right)")
cat(paste0("<center>", fig1ref, "</center>\n"))
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("fastq_len", "Read length distribution after adapter trimming", trunk.eval=has_fastq_len), eval=has_fastq_len}
cat("## Read length distribution after adapter trimming\n")
check_and_include_graphics(files$fastq_len)
```

# Host genome

## Overall 

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("category_mapped_bar", "Read distribution", trunk.eval=has_category_mapped_bar), eval=has_category_mapped_bar}
check_and_include_graphics(files$category_mapped_bar)
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("category_mapped_group", "Read distribution in grouped samples", trunk.eval=has_file(files,"category_mapped_group")),eval=has_file(files,"category_mapped_group")}
check_and_include_graphics(files$category_mapped_group)
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("category_mapped_bar", "Read distribution", trunk.eval=has_file(files,"category_table")), eval=has_file(files,"category_table")}
tbl = read.csv(files$category_table, header=TRUE, row.names=1, stringsAsFactor=FALSE)
ttbl = data.frame(t(tbl))
print(kable(ttbl, caption=tabRef("categoryTable", "Host genome category table"), row.names=TRUE) %>%
        kable_styling() %>%
        htmltools::HTML())
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("category_smallrna_bar", "smallRNA read distribution", trunk.eval=has_file(files,"category_smallrna_bar")), eval=has_file(files,"category_smallrna_bar")}
check_and_include_graphics(files$category_smallrna_bar)
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("category_smallrna_group", "smallRNA read distribution in grouped samples", trunk.eval=has_file(files,"category_smallrna_group")), eval=has_file(files,"category_smallrna_group")}
check_and_include_graphics(files$category_smallrna_group)
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("host_length_dist_category", "Host read length distribution in categories", trunk.eval=has_file(files,"host_length_dist_category")), eval=has_file(files,"host_length_dist_category")}
check_and_include_graphics(files$host_length_dist_category)
```

<br>

```{r hostgenome, child=hostGenomeFile} 
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("deseq2_host_vis", "DESeq2 result of host smallRNA", trunk.eval=hasHostDEseq2), eval=hasHostDEseq2}
cat('## host smallRNA DESeq2 result\n') 
if ("DE_use_raw_pvalue" %in% names(option_map) & option_map$DE_use_raw_pvalue == "1") {
  pValueText="p value"
} else {
  pValueText="FDR-adjusted p-value"
}
deseq2criteria<-paste0("absolute fold change >= ", option_map$DE_fold_change, " and ", pValueText, " <= ", option_map$DE_pvalue)
cat("Differential expression analysis criteria:", deseq2criteria, "\n\n")

check_and_include_graphics(files$deseq2_host_vis)
```

```{r, echo=FALSE, results='asis'}
if(hasNonhostGenome | hasNonhostLibrary){
  cat("# Non-Host")
}

```
<br>

```{r nonhostgenome, eval=hasNonhostGenome, child=nonhostGenomeFile} 
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("deseq2_nonhost_vis", "DESeq2 result of nonhost genomes", trunk.eval=hasNonhostDEseq2), eval=hasNonhostDEseq2}
cat('### Non-host genome DESeq2 result\n') 
if ("DE_use_raw_pvalue" %in% names(option_map) & option_map$DE_use_raw_pvalue == "1") {
  pValueText="p value"
} else {
  pValueText="FDR-adjusted p-value"
}
deseq2criteria<-paste0("absolute fold change >= ", option_map$DE_fold_change, " and ", pValueText, " <= ", option_map$DE_pvalue)
cat("Differential expression analysis criteria:", deseq2criteria, "\n\n")

check_and_include_graphics(files$deseq2_nonhost_vis)
```

<br>

```{r nonhostlibrary, eval=hasNonhostLibrary, child=nonhostLibraryFile} 
```

```{r bacteria_count_vis,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("bacteria_count_vis", "Bacteria count", trunk.eval=has_file(files,"bacteria_count_vis")), eval=has_file(files,"bacteria_count_vis")}
cat('## Non-Host bacteria count\n')

check_and_include_graphics(files$bacteria_count_vis)
```

```{r, echo=FALSE, results='asis'}
if(hasNonhostGenome | hasNonhostLibrary){
  cat("## Non-host summary")
}
```

```{r,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("nonhost_overlap_bar", "Read overlapped in non-host genome/database", trunk.eval=has_file(files,"nonhost_overlap_bar")), eval=has_file(files,"nonhost_overlap_bar")}
check_and_include_graphics(files$nonhost_overlap_bar)
```

```{r, echo=FALSE, results='asis'}
if(hasRefBacteriaEst){
  cat("# Refseq bacteria")
}
```

```{r refBacteriaAgg, eval=hasRefBacteriaAgg, child=nonhostBacteriaAggFile} 
```

```{r refBacteriaEst, eval=hasRefBacteriaEst, child=nonhostBacteriaEstFile} 
```

```{r teardown, include=FALSE}
if(file.exists(hostGenomeFile)){
  file.remove(hostGenomeFile)
}
if(file.exists(nonhostGenomeFile)){
  file.remove(nonhostGenomeFile)
}
if(file.exists(nonhostLibraryFile)){
  file.remove(nonhostLibraryFile)
}
if(file.exists(nonhostBacteriaAggFile)){
  file.remove(nonhostBacteriaAggFile)
}
if(file.exists(nonhostBacteriaEstFile)){
  file.remove(nonhostBacteriaEstFile)
}
```
# Software versions
```{r softwareVersion, echo=FALSE, results='asis'}
versions<-display_versions()
versions<-versions[!duplicated(versions$Software),]
```

# Description
```{r echo=FALSE}
vmap<-split(versions$Version, versions$Software)
vmap<-lapply(vmap, function(x) paste0(x, collapse=", "))

normalize_by_read<-option_map['normalize_by']
if(normalize_by_read == "FeatureReads"){
  if(!hasNonhostGenome){
    normalize_by_read = "total host smallRNA reads"
  }else{
    normalize_by_read = "total feature reads"
  }
}else{
  normalize_by_read = "total reads"
}

get_version<-function(name){
  if(name %in% names(vmap)){
    return(vmap[[name]])
  }else{
    return("")
  }
}
```

TIGER (“Tools for Integrative Genome analysis of Extracellular sRNAs”), an in-house small RNA sequencing analysis pipeline, was used to quantify small RNAs[1]. 
Briefly, Cutadapt (`r get_version('Cutadapt')`)[2] was used to trim 3’ adapters for raw reads. 
All reads with less than 16 nucleotides were designated as “too short” and discarded. 
Quality control on both raw reads and adaptor-trimmed reads was performed using FastQC (`r get_version('FastQC')`)(www.bioinformatics.babraham.ac.uk/projects/fastqc). 
The adaptor-trimmed reads were mapped to the `r get_version('host')` genome, addition to rRNA and tRNA reference sequences, 
by Bowtie1 (`r get_version('bowtie')`)[3] allowing only one mismatch. 
`r if(hasHostDEseq2){paste0("Significantly differential expressed features with ", deseq2criteria, " were detected by DESeq2 (", get_version('DESeq2'), ")[4] using ", normalize_by_read, " as normalization factor.")}` 
`r if(hasNonhostGenome){"The reads with less than 20 nucleotides which were neither annotated as small RNA nor perfectly mapped to human genome were discarded. 
Remaining unmapped reads were then mapped in parallel to exogenous structural RNA databases and curated 
microbial genome databases allowing no mismatches. Reads that failed to align to any strategy were categorized 
as 'unknown'."}` 

[1] Allen, R. M., Zhao, S., Ramirez Solano, M. A., Zhu, W., et al., Bioinformatic analysis of endogenous and exogenous small RNAs on lipoproteins. J Extracell Vesicles 2018, 7, 1506198. \
[2] Martin, M., Cutadapt removes adapter sequences from high-throughput sequencing reads. EMBnet.journal 2011, 17, 10-12. \
[3] Langmead, B., Trapnell, C., Pop, M., Salzberg, S. L., Ultrafast and memory-efficient alignment of short DNA sequences to the human genome. Genome Biol 2009, 10, R25. \
`r if(hasHostDEseq2){"[4] Love, M. I., Huber, W., Anders, S., Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biol 2014, 15, 550."}`
