---
title: "Whole Genome Bisulfite Sequencing Data Analysis Report"
date: "`r format(Sys.time())`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    number_sections: true
---

```{css, echo=FALSE}
#content{
    max-width:1920px;
}
```

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
library(tidyverse)
library(data.table)
library(stringr)
library(ggpubr)
library(purrr)
library(tibble)

source("Pipeline.R")
source("reportFunctions.R")

knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 120),
  message = FALSE,
  warning = FALSE,
  echo=FALSE,
  results='asis', 
  fig.align="center"
)

option_tbl<-read.table("fileList2.txt", sep="\t", header=FALSE, as.is=TRUE)
myoptions<-split(option_tbl$V1, option_tbl$V2)

project <- myoptions$task_name
meta <- myoptions$meta_data
abismal_path <- myoptions$abismal_path
dnmtools_path <- myoptions$dnmtools_path
MethylKitCorr_path <- myoptions$MethylKitCorr_path
MethylKitDiff_path <- myoptions$MethylKitDiff_path

files<-read.table("fileList1.txt", sep="\t", header=FALSE, as.is=TRUE)
rownames(files)<-files$V2

hasFastqc<-any(grepl("fastqc_", rownames(files)))

hasWebGestaltAnalysis=file.exists('fileList4.txt')

```

<br>

# Sequencing Quality
  
## Summary of sequencing quality
  
```{r fastqc_raw_per_base_sequence_quality, eval = hasFastqc, fig.cap=figRef("fastqc_raw_per_base_sequence_quality", "The mean quality scores of sequencing reads in each position", trunk.eval=file.exists(files["fastqc_raw_per_base_sequence_quality",1]))}
check_and_include_graphics(files["fastqc_raw_per_base_sequence_quality",1])
```

<br>

```{r fastqc_raw_per_base_sequence_quality_table, eval = hasFastqc}

tableFile=gsub(".png$","",files["fastqc_raw_per_base_sequence_quality",1])
table<-read.delim(tableFile)

QualityMin=30
if (any(table$Mean<=QualityMin)) { #use this as cutoff to show the quality table
  output_table1=data.frame(table[,c("File","Mean","Base")])
  output_table1$Base=factor(as.character(output_table1$Base),levels = unique(output_table1$Base))
  output_table1=tidyr::pivot_wider(output_table1,names_from=Base,values_from=Mean)
  output_table1=output_table1[which(apply(output_table1[,-1],1,min)< QualityMin),]
  
  print(kable(output_table1, caption=tabRef("Quality Summary", "Samples with Low WGBS Quality"), row.names=F) %>% 
          kable_styling() %>%
          row_spec(which(apply(output_table1[,-1],1,min)< QualityMin), color = "black", background = "honeydew") %>%
          htmltools::HTML())
}


```

<br>

```{r fastqc_raw_per_sequence_gc_content, eval = hasFastqc,  fig.cap=figRef("fastqc_raw_per_sequence_gc_content", "The average GC content of sequencing reads", trunk.eval=file.exists(files["fastqc_raw_per_sequence_gc_content",1]))}
check_and_include_graphics(files["fastqc_raw_per_sequence_gc_content",1])
```

<br>
  
```{r fastqc_raw_adapter_content,results='asis', eval = hasFastqc, fig.cap=figRef("fastqc_raw_adapter_content", "Adapter content of sequencing reads", trunk.eval=file.exists(files["fastqc_raw_adapter_content",1]))}
check_and_include_graphics(files["fastqc_raw_adapter_content",1])
```

<br>

# Mapping quality

## Summary of mapping quality

```{r mapping_summary}
files <- read.table('fileList5.txt', sep="\t")
pat = ".mapstats"
files = files[grepl(pat, files$V1),]

DT <- list()
i=1
for (i in 1:nrow(files) ) {
  sample = files$V2[i]
  infile <- files$V1[i]
  info <- read.delim(infile, header = F, sep = "", stringsAsFactors = FALSE, comment.char = "#", blank.lines.skip = TRUE, as.is = T)
  info <- info %>%
    dplyr::slice(2, 4:9, 11:17) %>%
    mutate(V1 = str_replace(V1, ":", "")) %>%
    column_to_rownames("V1")
  info <- t(info)
  rownames(info) <- sample
  DT[[sample]] <- info
}

mapping_df <- Reduce(rbind, DT)
mapping_df <- as.data.frame(mapping_df) %>% rownames_to_column("sample")
write.csv(file=paste0(project, "/", project, ".abismal.mapping.csv"), mapping_df)

mfinal <- mapping_df[,c("sample", "num_unique", "num_ambiguous", "num_unmapped")]
colnames(mfinal) <- c("Sample", "Unique", "Ambiguous", "Unmapped")
mfinal<-melt(mfinal, id="Sample", variable.name="Read", value.name="Count")
mfinal$Read<-factor(mfinal$Read, levels=c("Unmapped", "Ambiguous", "Unique"))

mapping_width=max(1500, 50 * ncol(mapping_df)) / 300
mapping_height=3

mapping_png=paste0(project, "/", project, ".abismal.mapping.csv.png")
g<-ggplot(mfinal, aes(x=Sample, y=Count, fill=Read)) +
  geom_bar(stat="identity", width=0.5) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, vjust=0.5, size=11, hjust=0, face="bold"),
        axis.text.y = element_text(size=11, face="bold"))
ggsave(mapping_png, g, width=mapping_width, height=mapping_height, units="in", dpi=300, bg="white")
```

```{r fig.width=mapping_width, fig.height=mapping_height, fig.cap=figRef("mapping_summary", "The statistics of WGBS mapping results")}
include_graphics(mapping_png)
```

<br>

```{r mapping_summary_table}
kable(mapping_df, caption="Mapping Quality Summary Table", row.names=F) %>% 
        kable_styling()
```

<br>


## Summary of coverage status

### Coverage proportion of Sites

```{r coverage_summary}
pat = ".levels"
files <- read.table('fileList6.txt', sep="\t")
files = files[grepl(pat, files$V1),]
samples=files$V2

DT <- list()
i=1
for (i in 1:nrow(files) ) {
  sample = files$V2[i]
  infile <- files$V1[i]
  info <- read.delim(infile, header = F, sep = " ", stringsAsFactors = FALSE, comment.char = "#", blank.lines.skip = TRUE, as.is = T)
  info <- info %>%
    mutate_at(c('V1'), ~na_if(., '')) %>%
    tidyr::fill(V1) %>%
    dplyr::select(-V2) %>%
    mutate(V1 = str_replace(V1, ":", ""),
           V3 = str_replace(V3, ":", "")) %>%
    dplyr::filter(!is.na(V4)) %>%
    dplyr::rename(Category = V1,
                  Stat = V3,
                  Value = V4)
  DT[[sample]] <- info$Value
}

levels_df <- Reduce(cbind, DT)
colnames(levels_df) <- samples
levels_df <- cbind(info[,c(1:2)], levels_df)

write.csv(file=paste0(project, "/", project, ".coverage.stat.csv"), levels_df)

mlevels <- levels_df %>%
  gather("Sample", "Value", 3:ncol(levels_df))

site_cov_perc <- mlevels %>%
  dplyr::filter(Stat == "sites_covered_fraction") %>%
  mutate(Category = factor(Category, levels = c("cpg", "cpg_symmetric", "cytosines", "chh", "cxg", "ccg")))

site_cov_perc_cpg <- site_cov_perc %>%
  dplyr::filter(Category == "cpg")

levels_width=max(1500, 50 * ncol(levels_df)) / 300
levels_height=3
levels_png=paste0(project, "/", project, ".coverage.stat.csv.png")

g1<-ggplot(site_cov_perc, aes(x=Sample, y=Value, fill=Category)) +
  geom_bar(stat="identity", width=0.5, position=position_dodge()) +
  xlab("") +
  ylab("Coverage proportion") +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, vjust=0.5, size=11, hjust=0.95, face="bold"),
        axis.text.y = element_text(size=11, face="bold"))
ggsave(levels_png, g1, width=levels_width, height=levels_height, units="in", dpi=300, bg="white")
```

```{r fig.width=levels_width, fig.height=levels_height, fig.cap=figRef("coverage_summary", "The coverage proportion of sites")}
include_graphics(levels_png)
```

<br>

### Coverage of target bases at different depths

```{r coverage_depth_summary}
files <- read.table('fileList5.txt', sep="\t")
pat = "hs_metrics.txt"
files = files[grepl(pat, files$V1),]
stopifnot(nrow(files) > 0)

DT <- list()
i=1
for (i in 1:nrow(files) ) {
  sample = files$V2[i]
  infile <- files$V1[i]
  info <- read.delim(infile, header = T, sep = "\t", stringsAsFactors = FALSE, comment.char = "#", blank.lines.skip = TRUE, nrows = 1)
  info$sample <- sample
  DT[[sample]] <- info
}

hs_met_df <- bind_rows(DT)
hs_met_df <- hs_met_df[, c(58, 1:54)]

write.csv(file=paste0(project, "/", project, ".HS.matrics.csv"), hs_met_df)

cov_perc_depth <- hs_met_df %>%
  dplyr::select(c("sample", starts_with("PCT_TARGET_BASES"))) %>%
  gather("Depth", "Percentage", -sample) %>%
  mutate(Depth = gsub("PCT_TARGET_BASES_(.+)", "\\1", Depth)) %>%
  mutate(Depth = factor(Depth, levels = c("1X", "2X", "10X", "20X", "30X", "40X", "50X", "100X")))

cov_width=max(1500, 50 * nrow(files)) / 300
cov_height=3
cov_png = paste0(project, "/", project, ".coverage.depth.perc.csv.png")

g2<-ggplot(cov_perc_depth, aes(x=sample, y=Percentage, group = Depth, colour=Depth)) +
  geom_line(size = 1.5) +
  geom_point() +
  xlab("") +
  ylab("Coverage percentage") +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, vjust=0.5, size=11, hjust=0.95, face="bold"),
        axis.text.y = element_text(size=11, face="bold"))
ggsave(cov_png, g2, width=cov_width, height=cov_height, units="in", dpi=300, bg="white")
```

```{r fig.width=cov_width, fig.height=cov_height, fig.cap=figRef("coverage_depth_summary", "The Coverage proportion of target bases at different depths")}
include_graphics(cov_png)
```

<br>

## Duplication rate statistics

```{r dup_summary}
pat <- ".bam.dupstats$"
files <- read.table('fileList5.txt', sep="\t")
files = files[grepl(pat, files$V1),]
stopifnot(nrow(files) > 0)
DT <- list()
i=1
for (i in 1:nrow(files) ) {
  sample = files$V2[i]
  infile <- files$V1[i]
  info <- read.delim(infile, header = F, sep = "", stringsAsFactors = FALSE, comment.char = "#", blank.lines.skip = TRUE, as.is = T)
  info <- info %>%
    mutate(V1 = str_replace(V1, ":", "")) %>%
    column_to_rownames("V1")
  info <- t(info)
  rownames(info) <- sample
  DT[[sample]] <- info
}

dup_df <- Reduce(rbind, DT)

write.csv(file=paste0(project, "/", project, ".dup.rate.csv"), dup_df)

dup_rate <- as.data.frame(dup_df) %>%
  rownames_to_column("Sample") %>%
  dplyr::select(c("Sample", "duplication_rate"))

dup_width=max(1500, 50 * nrow(files)) / 300
dup_height=3

dup_png=paste0(project, "/", project, ".dup.rate.csv.png")

g3<-ggplot(dup_rate, aes(x=Sample, y=duplication_rate)) +
  geom_bar(stat="identity", width = 0.5, fill = "blue") +
  ylim(0,100) +
  xlab("") +
  ylab("Duplication Rate") +
  theme_classic() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle=90, vjust=0.5, size=11, hjust=0.95, face="bold"),
        axis.text.y = element_text(size=11, face="bold"))
ggsave(dup_png, g3, width=dup_width, height=dup_height, units="in", dpi=300, bg="white")
```

```{r fig.width=dup_width, fig.height=dup_height, fig.cap=figRef("dup_summary", "The duplication rate")}
include_graphics(dup_png)
```

<br>

# Methylation statistics

## BS convert rate statistics 

### The overall BS rate

```{r bs_summary}
pat <- ".bsrate$"
files <- read.table('fileList6.txt', sep="\t")
files = files[grepl(pat, files$V1),]
stopifnot(nrow(files) > 0)
samples=files$V2
bsrate <- c()
i=1
for (i in 1:nrow(files) ) {
  sample = files$V2[i]
  infile <- files$V1[i]
  info <- read.table(infile, header = F, sep = "", stringsAsFactors = FALSE, comment.char = "#", blank.lines.skip = TRUE, as.is = T, nrows = 1)
  info <- info$V5
  bsrate <- append(bsrate, info)
}

bs_overall_df <- data.frame(samples, bsrate)

write.csv(file=paste0(project, "/", project, ".bs.overall.rate.csv"), bs_overall_df)

bs_width=max(1500, 50 * nrow(files)) / 300
bs_height=3
bs_png=paste0(project, "/", project, ".bs.overall.rate.csv.png")

g4<-ggplot(bs_overall_df, aes(x=samples, y=bsrate)) +
  geom_bar(stat="identity", width = 0.5, fill = "green") +
  xlab("") +
  ylab("BS Rate") +
  theme_classic() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle=90, vjust=0.5, size=11, hjust=0.95, face="bold"),
        axis.text.y = element_text(size=11, face="bold"))
ggsave(bs_png, g4, width=bs_width, height=bs_height, units="in", dpi=300, bg="white")
```

```{r fig.width=bs_width, fig.height=bs_height, fig.cap=figRef("bs_summary", "The BS convert rate")}
include_graphics(bs_png)
```

<br>

### The BS rates of Non-CpG and CpG sites

```{r bs_cpg_summary, fig.cap=figRef("bs_cpg_summary", "The BS convert rates of Non-CpG and CpG sites")}
pat <- ".rrbs_summary_metrics$"
files <- read.table('fileList5.txt', sep="\t")
files = files[grepl(pat, files$V1),]
stopifnot(nrow(files) > 0)

DT <- list()
i=1
for (i in 1:nrow(files) ) {
  sample = files$V2[i]
  infile <- files$V1[i]
  info <- read.delim(infile, header = T, sep = "\t", stringsAsFactors = FALSE, comment.char = "#", blank.lines.skip = TRUE, nrows = 1)
  rownames(info) <- sample
  DT[[sample]] <- info
}

rrbs_met_df <- bind_rows(DT)

write.csv(file=paste0(project, "/", project, ".bs.CpG.rate.csv"), bs_overall_df)

rrbs_final <- rrbs_met_df %>%
  rownames_to_column("Sample") %>%
  dplyr::select(Sample, PCT_NON_CPG_BASES_CONVERTED, PCT_CPG_BASES_CONVERTED) %>%
  gather("Site_type", "Percentage", -Sample) %>%
  mutate(Percentage = ifelse(Percentage <= 1, Percentage, 1-(Percentage - 1)))

rate_width=max(1500, 50 * nrow(files)) / 300 + 2
rate_height=3
rate_png=paste0(project, "/", project, ".bs.CpG.rate.csv.png")

g5<-ggplot(rrbs_final, aes(x=Sample, y=Percentage, fill = Site_type)) +
  geom_bar(stat="identity", position=position_dodge()) +
  xlab("") +
  ylab("BS rate") +
  ylim(0,1) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, vjust=0.5, size=11, hjust=0.95, face="bold"),
        axis.text.y = element_text(size=11, face="bold"))
ggsave(rate_png, g5, width=rate_width, height=rate_height, units="in", dpi=300, bg="white")
```

```{r fig.width=rate_width, fig.height=rate_height, fig.cap=figRef("bs_cpg_summary", "The BS convert rates of Non-CpG and CpG sites")}
include_graphics(rate_png)
```
<br>

## The statistics of methylation level

### the fraction of "called" sites "called" methylated

```{r methy_summary1, fig.cap=figRef("methy_summary1", "The fraction of methylated sites")}
fractional_meth <- mlevels %>%
  dplyr::filter(Stat == "fractional_meth") %>%
  mutate(Category = factor(Category, levels = c("cpg", "cpg_symmetric", "cytosines", "chh", "cxg", "ccg"))) %>%
  dplyr::filter(Category == "cpg")

nsample = length(unique(fractional_meth$Sample))

fra_width=max(1500, 50 * nsample) / 300
fra_height=3
fra_png=paste0(project, "/", project, ".fractional_meth.csv.png")

g6<-ggplot(fractional_meth, aes(x=Sample, y=Value, fill=Category)) +
  geom_bar(stat="identity", width=0.5, position=position_dodge()) +
  xlab("") +
  ylab("The fraction of methylated sites") +
  ylim(0,1) +
  theme_classic() +
  theme(legend.position="none",
        axis.text.x = element_text(angle=90, vjust=0.5, size=11, hjust=0.95, face="bold"),
        axis.text.y = element_text(size=11, face="bold"))

ggsave(fra_png, g6, width=fra_width, height=fra_height, units="in", dpi=300, bg="white")
```

```{r fig.width=fra_width, fig.height=fra_height, fig.cap=figRef("methy_summary1", "The fraction of methylated sites")}
include_graphics(fra_png)
```

<br>

### the mean of the methylation levels for covered sites

```{r methy_summary2, fig.cap=figRef("methy_summary2", "The mean of the methylation levels")}
mean_meth <- mlevels %>%
  dplyr::filter(Stat == "mean_meth") %>%
  mutate(Category = factor(Category, levels = c("cpg", "cpg_symmetric", "cytosines", "chh", "cxg", "ccg"))) %>%
  dplyr::filter(Category == "cpg")

nsample = length(unique(mean_meth$Sample))

mean_width=max(1500, 50 * nsample) / 300
mean_height=3
mean_png=paste0(project, "/", project, ".mean_meth.csv.png")

g7 <- ggplot(mean_meth, aes(x=Sample, y=Value, fill=Category)) +
  geom_bar(stat="identity", width=0.5, position=position_dodge()) +
  xlab("") +
  ylab("The mean of the methylation levels") +
  ylim(0,1) +
  theme_classic() +
  theme(legend.position="none",
        axis.text.x = element_text(angle=90, vjust=0.5, size=11, hjust=0.95, face="bold"),
        axis.text.y = element_text(size=11, face="bold"))
ggsave(mean_png, g7, width=mean_width, height=mean_height, units="in", dpi=300, bg="white")
```

```{r fig.width=mean_width, fig.height=mean_height, fig.cap=figRef("methy_summary2", "The mean of the methylation levels")}
include_graphics(mean_png)
```

<br>

### the mean of the methylation levels weighted by coverage

```{r methy_summary3, fig.cap=figRef("methy_summary3", "The mean of the methylation levels weighted by coverage")}
mean_meth_weighted <- mlevels %>%
  dplyr::filter(Stat == "mean_meth_weighted") %>%
  mutate(Category = factor(Category, levels = c("cpg", "cpg_symmetric", "cytosines", "chh", "cxg", "ccg"))) %>%
  dplyr::filter(Category == "cpg")

nsample = length(unique(mean_meth_weighted$Sample))

mean_width=max(1500, 50 * nsample) / 300
mean_height=3
mean_png=paste0(project, "/", project, ".mean_meth_weighted.csv.png")

g8 <- ggplot(mean_meth_weighted, aes(x=Sample, y=Value, fill=Category)) +
  geom_bar(stat="identity", width=0.5, position=position_dodge()) +
  xlab("") +
  ylab("mean_meth_weighted") +
  ylim(0,1) +
  theme_classic() +
  theme(legend.position="none",
        axis.text.x = element_text(angle=90, vjust=0.5, size=11, hjust=0.95, face="bold"),
        axis.text.y = element_text(size=11, face="bold"))

ggsave(mean_png, g8, width=mean_width, height=mean_height, units="in", dpi=300, bg="white")
```

```{r fig.width=mean_width, fig.height=mean_height, fig.cap=figRef("methy_summary3", "The mean of the methylation levels weighted by coverage")}
include_graphics(mean_png)
```

<br>

# correlation distance of β value of CpG sites


```{r dmcpg_table, include=TRUE, out.height = "680px", out.width='800px', fig.cap=c("MDS plot based on correlation distance of β value of CpG sites")}
knitr::include_graphics(paste0(MethylKitCorr_path, project, "_methyl_CpG_bvalue_corr_MDS_plot.png"))

#mincov <- 4
#cpg.pat <- ".meth$"
#
#cpg.all <- list.files(path = dnmtools_path, pattern = cpg.pat, all.files = FALSE, recursive = T, ignore.case = FALSE, include.dirs = FALSE)
#cpg.all <- cpg.all[!(str_detect(cpg.all,"all.meth$"))]
#cov.lst <- list()
#cpg.lst <- list()
#for (i in 1:length(cpg.all)) {
#	#read in *.meth files
#	cpg.infile <- paste(dnmtools_path, cpg.all[i], sep = "/")
#	id <- gsub("(.*?)\\.meth$", "\\1", cpg.all[i])
#	#format cpg data frame
#	DT.cpg <- read.table(cpg.infile, header = F, stringsAsFactors = FALSE)
#	colnames(DT.cpg) <- c("chr", "base", "strand", "Type", id, "coverage")
#	DT.cpg$chrBase <- paste(DT.cpg$chr, DT.cpg$base, sep = ".")
#	DT.cov <- DT.cpg[,c("chrBase", "coverage")]
#	colnames(DT.cov) <- c("chrBase", id)
#  DT.cpg <- DT.cpg[,c("chrBase", id)]
#  cov.lst[[id]] <- DT.cov
#	cpg.lst[[id]] <- DT.cpg
#}
#
#cov.all.df <- cov.lst %>% 
#    purrr::reduce(full_join, by='chrBase') %>%
#    mutate_at(names(cov.lst), as.numeric) %>%
#    dplyr::filter(if_any(is.numeric, ~ .x > mincov))
#left_sites <- cov.all.df$chrBase
#  
#cpg.all.df <- cpg.lst %>% 
#  purrr::reduce(full_join, by='chrBase') %>%
#  mutate_at(names(cpg.lst), as.numeric) %>%
#  dplyr::filter(chrBase %in% left_sites) %>%
#  column_to_rownames("chrBase")
#saveRDS(cpg.all.df, file = paste0(project, "/", project, ".all.cpg.mincov", mincov, ".rds"))
#
#cpg_bvalue_cor <- cor(cpg.all.df, method = "pearson", use = "complete.obs")
#
#cpg_bvalue_mds <- (1 - cpg_bvalue_cor) %>%
#  cmdscale() %>%
#  as_tibble()
#
#colnames(cpg_bvalue_mds) <- c("Dim.1", "Dim.2")
##rownames(cpg_bvalue_mds) <- colnames(cpg_bvalue_df)
#meta_df <- read.table(meta, sep = "\t", header = T)
##cpg_bvalue_mds$group <- c(rep("DKs8", 4), rep("DLD1", 5), rep("Kidney", 3))
##cpg_bvalue_mds$type <- c("cell", "EV", "EV", "EV", "cell", "EV", "EV", "EV", "EV", "EV", "cell", "EV")
#cpg_bvalue_mds$group <- meta_df$group
#cpg_bvalue_mds$type <- meta_df$type
#
## Plot MDS
#dms_plot <- ggscatter(cpg_bvalue_mds, x = "Dim.1", y = "Dim.2", 
#                      label = colnames(cpg.all.df),
#                      xlab = "MDS 1",
#                      ylab = "MDS 2",
#                      color = "group",
#                      palette = "jco",
#                      size = 1, 
#                      ellipse = F,
#                      ellipse.type = "norm",
#                      repel = TRUE) + 
#  theme_bw()
#
#
#pdf(file = paste0(project, "/", project, ".methyl_CpG_bvalue_cor_MDS_plot.pdf"), width = 6, height = 6, pointsize = 8, onefile = T)
#dms_plot
#invisible(dev.off())
#
#png(file=paste0(project, "/", project, ".methyl_CpG_bvalue_cor_MDS_plot.png"), height=1500, width=1500, res=300)
#print(dms_plot)
#invisible(dev.off())
#
#dms_plot
```


<br>

# differentially methylated CpGs


```{r mds, include=TRUE, message = F, warning = F}
dmcpg.pat <- ".dmcpgs$"
files <- read.table("fileList8.txt", sep="\t", header=FALSE, stringsAsFactors = FALSE)

dmcpg.lst <- list()
i=1
for (i in 1:nrow(files)) {
	#read in *.meth files
  comparison_group=files$V2[i]
	infile=files$V1[i]

	comparison <- basename(dirname(infile))
  group = gsub(paste0("^", comparison, "_"), "", comparison_group) 
	
	#format cpg data frame
	DT.dmcpg <- read.table(infile, header = T, sep = "\t", stringsAsFactors = FALSE)
	hypo_CpGs_num <- nrow(DT.dmcpg)
	link <- paste0("[",group,"](./", project, "/", comparison_group, ".dmcpgs)")
	info <- c(comparison, group, hypo_CpGs_num, link)
	dmcpg.lst[[i]] <- info
}

dmcpg_df <- Reduce(rbind, dmcpg.lst)
colnames(dmcpg_df) <- c("Comparison", "Hypo_in_Group", "Hypo_meth_num", "Links")

write.table(paste0(project, "/", project, "_dmcpg_sign_file.txt"), sep = "\t", quote = F)

kable(dmcpg_df, caption="Links of differentially methylated CpGs for comparisons", row.names=F) %>% 
        kable_styling()
```

<br>

```{r include=FALSE, eval=hasWebGestaltAnalysis}
wga_rmd = "webGestaltAnalysis.rmd"

wga_files=read.table('fileList4.txt', sep="\t", header=FALSE, stringsAsFactors = FALSE)
wga_files=wga_files[str_detect(wga_files$V1, "txt$"),]
wga_names=unique(wga_files$V2)

fdr_cutoff <- 0.1

result = "# Enrichment of cis-regulated genes of differentially methylated CpGs\n<br>\n\n"
wga_name = wga_names[1]
for(wga_name in wga_names){
  result = paste0(result, paste0("## ", wga_name, "\n\n"))
  cur_tbl = wga_files[wga_files$V2 == wga_name,]
  cur_file = cur_tbl$V1[1]
  prefix = paste0("enrichment_results_", wga_name)
  for(cur_file in cur_tbl$V1){
    if(file.exists(cur_file)){
      cur_name = basename(cur_file)
      cur_category = str_sub(cur_name, nchar(prefix) +2, -5)
      result = paste0(result, paste0("### ", cur_category, "\n\n"))
      result = paste0(result, '```{r include=T, warning = F, message = F, echo = F}\nanno_df <- read.table("', cur_file, '", sep = "\\t", header = T)\n')
      result = paste0(result, 'anno_df <- anno_df %>% dplyr::select(-c(link, overlapId)) %>% dplyr::filter(FDR <= fdr_cutoff)\n')
      result = paste0(result, 'anno_df %>% DT::datatable(extensions = "Buttons", options = list(dom = "Bfrtip", buttons = c("excel", "csv")))\n```\n\n')
    }
  }
}

cat(result, file=wga_rmd)
```

```{r, child=wga_rmd, eval=hasWebGestaltAnalysis} 
```

<br>
